<a href="https://colab.research.google.com/github/JoshuaFiallos/SearchAlgorithms/blob/master/BSF_DFS_Astar_p54.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Integrantes**

Luis Andy Briones Peñafiel

Manuel Agustin Aycart Armijos

Alexander Jose Jimenez Gurumendi

Hernan Alejandro Rodriguez Veliz

Luis Antonio Palomino Romero

Jesus David Holguin Mendoza

RICARDO ANDRES AUCAPINA BRIONES

Steven Xavier Sanchez Muñoz

Danny Jacinto Figueroa Smith

Luis Andres Rodriguez Morales

Ricardo Escalante 

DANIEL ENRIQUE VEGA BRAVO





#**Metodos de busqueda BFS,DFS y A***

**GRAFO PARA DFS, BFS y A***     


Un grafo consiste en un conjunto de nodos y un conjunto de arcos dirigidos entre nodos. La idea es encontrar una ruta a lo largo de estos arcos desde el nodo de inicio a un nodo objetivo:


![texto alternativo](https://i2.wp.com/aiandgames.com/wp-content/uploads/2014/09/romania-graph1.png)




Se realiza un grafo para aquello usaremos la lista de adyacencia. Un método efectivo / elegante para implementar listas de adyacencia en Python es usar diccionarios. Las claves del diccionario representan nodos, los valores tienen una lista de vecinos

In [0]:
graph = {'ARAD': set(['ZERIND', 'SIBIU','TIMISOARA']),
         'BUCHAREST': set(['FAGARAS', 'PITESTI','URZICENI','GIURGIU']),
         'CRAIOVA': set(['RIMNICU VILCEA', 'DOBRETA','PITESTI']),
         'DOBRETA': set(['MEHADIA', 'CRAIOVA']),
         'EFORIE': set(['HIRSOVA']),
         'FAGARAS': set(['SIBIU', 'BUCHAREST']),
         'GIURGIU': set(['BUCHAREST']),
         'HIRSOVA': set(['URZICENI', 'EFORIE']),
         'IASI': set(['NEAMT', 'VASLUI']),
         'LUGOJ': set(['TIMISOARA', 'MEHADIA']),
         'MEHADIA': set(['LUGOJ', 'DOBRETA']),
         'NEAMT': set(['IASI']),
         'ORADEA': set(['ZERIND', 'SIBIU']),
         'PITESTI': set(['RIMNICU VILCEA', 'CRAIOVA','BUCHAREST']),
         'RIMNICU VILCEA': set(['PITESTI', 'SIBIU','CRAIOVA']),
         'SIBIU': set(['ORADEA','ARAD','FAGARAS','RIMNICU VILCEA']),
         'TIMISOARA': set(['ARAD', 'LUGOJ']),
         'URZICENI': set(['BUCHAREST', 'HIRSOVA','VASLUI']),
         'VASLUI': set(['IASI', 'URZICENI']),
         'ZERIND': set(['ORADEA', 'ARAD'])}

# ALGORITMO BFS

Se realizó la función BDF para recorrer un grafo y encontrar el camino más corto. Primero, en el caso de la aplicación de la ruta más corta, necesitamos que queue mantenga un registro de las rutas posibles (implementadas como lista de nodos) en lugar de nodos. Segundo, cuando el algoritmo verifica si hay un nodo vecino, debe verificar si el nodo vecino corresponde al nodo objetivo. Si ese es el caso, tenemos una solución y no hay necesidad de seguir explorando el gráfico

In [0]:
# encuentra la ruta más corta entre 2 nodos de un gráfico usando BFS
def bfs_shortest_path(graph, start, goal):
  
    print('Nodo inicial: ',start)
    print ('Nodo objetivo: ',goal)
    print('\n')

    
    # hacer un seguimiento de los nodos explorados
    explored = []
    # realizar un seguimiento de todos los caminos que deben comprobarse
    queue = [[start]]
    
    # retornar la ruta si el inicio es el objetivo
    if start == goal:
        return "Eso fue fácil! Inicio = objetivo"
 
    # mantenernos en bucle hasta que se hayan verificado todos los caminos posibles
    while queue:
        # Pop el primer camino de la cola
        path = queue.pop(0)
        
        # obtener el último nodo de la ruta
        node = path[-1]
        print('Ruta en la cola: ',path, ' - Último nodo de la ruta ',node)
        if node not in explored:
            print(f'Nodo no explorado {node} ')
            neighbours = graph[node]
            # Ir a través de todos los nodos vecinos, construir una nueva ruta y empujarlo en la cola
            print('Nodos vecinos de : ',node,' son: ',neighbours)
            for neighbour in neighbours:
                new_path = list(path)                
                new_path.append(neighbour)
                queue.append(new_path)
                print('Nuevas Rutas', new_path)
                
                # retornar ruta si el vecino es el objetivo
                if neighbour == goal:
                  return 'Camino Escogido: ', new_path
            print('No se encontro Objetivo')
            print('Nuevas rutas agregada a la cola: ',queue)
            print('-------------------------------------------------------------------------------------------------------------')

            # marca el nodo como explorado
            explored.append(node)
        else:
            print(f'Ya ha sido explorado nodo: {node}')
 
    # En caso de que no haya ruta entre los 2 nodos.
    return "Lo siento, pero no existe una ruta de conexión..."

**PRUEBA 1 BFS**

En la  prueba 1 se puede observar que el nodo de inicio es ARAD recorre el grafo en anchura y encuentra el camino más corto tal que queue va registrando las rutas posibles en lugar de nodos. Cuando el algoritmo verifica si hay un nodo vecino, debe verificar si el nodo vecino corresponde al nodo objetivo. Esta prueba el objetivo de SIBIU si se encuentra ya no hay necesidad de seguir explorando el gráfo.

In [0]:
bfs_shortest_path(graph,'ARAD','SIBIU')

Nodo inicial:  ARAD
Nodo objetivo:  SIBIU


Ruta en la cola:  ['ARAD']  - Último nodo de la ruta  ARAD
Nodo no explorado ARAD 
Nodos vecinos de :  ARAD  son:  {'ZERIND', 'SIBIU', 'TIMISOARA'}
Nuevas Rutas ['ARAD', 'ZERIND']
Nuevas Rutas ['ARAD', 'SIBIU']


('Camino Escogido: ', ['ARAD', 'SIBIU'])

**PRUEBA 2 BFS**

En la prueba 2 se puede observar que el nodo de inicio es PISTESTI recorre el grafo en anchura y encuentra el camino más corto tal que queue va registrando las rutas posibles en lugar de nodos. Cuando el algoritmo verifica si hay un nodo vecino, debe verificar si el nodo vecino corresponde al nodo objetivo. Esta prueba el objetivo de EFORIE si se encuentra ya no hay necesidad de seguir explorando el gráfo.

In [0]:
bfs_shortest_path(graph,'PITESTI','EFORIE')

Nodo inicial:  PITESTI
Nodo objetivo:  EFORIE


Ruta en la cola:  ['PITESTI']  - Último nodo de la ruta  PITESTI
Nodo no explorado PITESTI 
Nodos vecinos de :  PITESTI  son:  {'CRAIOVA', 'BUCHAREST', 'RIMNICU VILCEA'}
Nuevas Rutas ['PITESTI', 'CRAIOVA']
Nuevas Rutas ['PITESTI', 'BUCHAREST']
Nuevas Rutas ['PITESTI', 'RIMNICU VILCEA']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['PITESTI', 'CRAIOVA'], ['PITESTI', 'BUCHAREST'], ['PITESTI', 'RIMNICU VILCEA']]
-------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['PITESTI', 'CRAIOVA']  - Último nodo de la ruta  CRAIOVA
Nodo no explorado CRAIOVA 
Nodos vecinos de :  CRAIOVA  son:  {'DOBRETA', 'RIMNICU VILCEA', 'PITESTI'}
Nuevas Rutas ['PITESTI', 'CRAIOVA', 'DOBRETA']
Nuevas Rutas ['PITESTI', 'CRAIOVA', 'RIMNICU VILCEA']
Nuevas Rutas ['PITESTI', 'CRAIOVA', 'PITESTI']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['PITESTI', 'BUCHAREST'], ['P

('Camino Escogido: ',
 ['PITESTI', 'BUCHAREST', 'URZICENI', 'HIRSOVA', 'EFORIE'])

**PRUEBA 3 BFS**


En la prueba 2 se puede observar que el nodo de inicio es SIBIU recorre el grafo en anchura y encuentra el camino más corto tal que queue va registrando las rutas posibles en lugar de nodos. Cuando el algoritmo verifica si hay un nodo vecino, debe verificar si el nodo vecino corresponde al nodo objetivo. Esta prueba el objetivo de ARAD si se encuentra ya no hay necesidad de seguir explorando el gráfo.

In [0]:
bfs_shortest_path(graph,'SIBIU','ARAD')

Nodo inicial:  SIBIU
Nodo objetivo:  ARAD


Ruta en la cola:  ['SIBIU']  - Último nodo de la ruta  SIBIU
Nodo no explorado SIBIU 
Nodos vecinos de :  SIBIU  son:  {'RIMNICU VILCEA', 'ORADEA', 'ARAD', 'FAGARAS'}
Nuevas Rutas ['SIBIU', 'RIMNICU VILCEA']
Nuevas Rutas ['SIBIU', 'ORADEA']
Nuevas Rutas ['SIBIU', 'ARAD']


('Camino Escogido: ', ['SIBIU', 'ARAD'])

**PRUEBA 4 BFS**

En la prueba 4 se puede observar que el nodo de inicio es SIBIU recorre el grafo en anchura y encuentra el camino más corto tal que queue va registrando las rutas posibles en lugar de nodos. Cuando el algoritmo verifica si hay un nodo vecino, debe verificar si el nodo vecino corresponde al nodo objetivo. Esta prueba el objetivo de BUCHAREST si se encuentra ya no hay necesidad de seguir explorando el gráfo.

In [0]:
bfs_shortest_path(graph,'SIBIU','BUCHAREST')

Nodo inicial:  SIBIU
Nodo objetivo:  BUCHAREST


Ruta en la cola:  ['SIBIU']  - Último nodo de la ruta  SIBIU
Nodo no explorado SIBIU 
Nodos vecinos de :  SIBIU  son:  {'RIMNICU VILCEA', 'ORADEA', 'ARAD', 'FAGARAS'}
Nuevas Rutas ['SIBIU', 'RIMNICU VILCEA']
Nuevas Rutas ['SIBIU', 'ORADEA']
Nuevas Rutas ['SIBIU', 'ARAD']
Nuevas Rutas ['SIBIU', 'FAGARAS']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['SIBIU', 'RIMNICU VILCEA'], ['SIBIU', 'ORADEA'], ['SIBIU', 'ARAD'], ['SIBIU', 'FAGARAS']]
-------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['SIBIU', 'RIMNICU VILCEA']  - Último nodo de la ruta  RIMNICU VILCEA
Nodo no explorado RIMNICU VILCEA 
Nodos vecinos de :  RIMNICU VILCEA  son:  {'SIBIU', 'CRAIOVA', 'PITESTI'}
Nuevas Rutas ['SIBIU', 'RIMNICU VILCEA', 'SIBIU']
Nuevas Rutas ['SIBIU', 'RIMNICU VILCEA', 'CRAIOVA']
Nuevas Rutas ['SIBIU', 'RIMNICU VILCEA', 'PITESTI']
No se encontro Objetivo
Nuevas ruta

('Camino Escogido: ', ['SIBIU', 'FAGARAS', 'BUCHAREST'])

**PRUEBA 5 BFS**

En la prueba 5 se puede observar que el nodo de inicio es ORADEA recorre el grafo en anchura y encuentra el camino más corto tal que queue va registrando las rutas posibles en lugar de nodos. Cuando el algoritmo verifica si hay un nodo vecino, debe verificar si el nodo vecino corresponde al nodo objetivo. Esta prueba el objetivo de CRAIOVA si se encuentra ya no hay necesidad de seguir explorando el gráfo.

In [0]:
bfs_shortest_path(graph,'ORADEA','CRAIOVA')

Nodo inicial:  ORADEA
Nodo objetivo:  CRAIOVA


Ruta en la cola:  ['ORADEA']  - Último nodo de la ruta  ORADEA
Nodo no explorado ORADEA 
Nodos vecinos de :  ORADEA  son:  {'ZERIND', 'SIBIU'}
Nuevas Rutas ['ORADEA', 'ZERIND']
Nuevas Rutas ['ORADEA', 'SIBIU']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['ORADEA', 'ZERIND'], ['ORADEA', 'SIBIU']]
-------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['ORADEA', 'ZERIND']  - Último nodo de la ruta  ZERIND
Nodo no explorado ZERIND 
Nodos vecinos de :  ZERIND  son:  {'ORADEA', 'ARAD'}
Nuevas Rutas ['ORADEA', 'ZERIND', 'ORADEA']
Nuevas Rutas ['ORADEA', 'ZERIND', 'ARAD']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['ORADEA', 'SIBIU'], ['ORADEA', 'ZERIND', 'ORADEA'], ['ORADEA', 'ZERIND', 'ARAD']]
-------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['ORADEA', 'SIBIU']  -

('Camino Escogido: ', ['ORADEA', 'SIBIU', 'RIMNICU VILCEA', 'CRAIOVA'])

**PRUEBA  6 BFS**

En la prueba 6 se puede observar que el nodo de inicio es PITESTI recorre el grafo en anchura y encuentra el camino más corto tal que queue va registrando las rutas posibles en lugar de nodos. Cuando el algoritmo verifica si hay un nodo vecino, debe verificar si el nodo vecino corresponde al nodo objetivo. Esta prueba el objetivo de URZICENI si se encuentra ya no hay necesidad de seguir explorando el gráfo.

In [0]:
bfs_shortest_path(graph,'PITESTI','URZICENI')

Nodo inicial:  PITESTI
Nodo objetivo:  URZICENI


Ruta en la cola:  ['PITESTI']  - Último nodo de la ruta  PITESTI
Nodo no explorado PITESTI 
Nodos vecinos de :  PITESTI  son:  {'CRAIOVA', 'BUCHAREST', 'RIMNICU VILCEA'}
Nuevas Rutas ['PITESTI', 'CRAIOVA']
Nuevas Rutas ['PITESTI', 'BUCHAREST']
Nuevas Rutas ['PITESTI', 'RIMNICU VILCEA']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['PITESTI', 'CRAIOVA'], ['PITESTI', 'BUCHAREST'], ['PITESTI', 'RIMNICU VILCEA']]
-------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['PITESTI', 'CRAIOVA']  - Último nodo de la ruta  CRAIOVA
Nodo no explorado CRAIOVA 
Nodos vecinos de :  CRAIOVA  son:  {'DOBRETA', 'RIMNICU VILCEA', 'PITESTI'}
Nuevas Rutas ['PITESTI', 'CRAIOVA', 'DOBRETA']
Nuevas Rutas ['PITESTI', 'CRAIOVA', 'RIMNICU VILCEA']
Nuevas Rutas ['PITESTI', 'CRAIOVA', 'PITESTI']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['PITESTI', 'BUCHAREST'], [

('Camino Escogido: ', ['PITESTI', 'BUCHAREST', 'URZICENI'])

**PRUEBA 7 BFS**


En la prueba 7 se puede observar que el nodo de inicio es NEAMT recorre el grafo en anchura y encuentra el camino más corto tal que queue va registrando las rutas posibles en lugar de nodos. Cuando el algoritmo verifica si hay un nodo vecino, debe verificar si el nodo vecino corresponde al nodo objetivo. Esta prueba el objetivo de URZICENI si se encuentra ya no hay necesidad de seguir explorando el gráfo.

In [0]:
bfs_shortest_path(graph,'NEAMT','URZICENI')

Nodo inicial:  NEAMT
Nodo objetivo:  URZICENI


Ruta en la cola:  ['NEAMT']  - Último nodo de la ruta  NEAMT
Nodo no explorado NEAMT 
Nodos vecinos de :  NEAMT  son:  {'IASI'}
Nuevas Rutas ['NEAMT', 'IASI']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['NEAMT', 'IASI']]
-------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['NEAMT', 'IASI']  - Último nodo de la ruta  IASI
Nodo no explorado IASI 
Nodos vecinos de :  IASI  son:  {'NEAMT', 'VASLUI'}
Nuevas Rutas ['NEAMT', 'IASI', 'NEAMT']
Nuevas Rutas ['NEAMT', 'IASI', 'VASLUI']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['NEAMT', 'IASI', 'NEAMT'], ['NEAMT', 'IASI', 'VASLUI']]
-------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['NEAMT', 'IASI', 'NEAMT']  - Último nodo de la ruta  NEAMT
Ya ha sido explorado nodo: NEAMT
Ruta en la cola:  ['NEAMT', 'IASI', 'VASLUI

('Camino Escogido: ', ['NEAMT', 'IASI', 'VASLUI', 'URZICENI'])

**PRUEBA 8 BFS**

En la prueba 8 se puede observar que el nodo de inicio es ORADEA recorre el grafo en anchura y encuentra el camino más corto tal que queue va registrando las rutas posibles en lugar de nodos. Cuando el algoritmo verifica si hay un nodo vecino, debe verificar si el nodo vecino corresponde al nodo objetivo. Esta prueba el objetivo de FAGARAS si se encuentra ya no hay necesidad de seguir explorando el gráfo.

In [0]:
bfs_shortest_path(graph,'ORADEA','FAGARAS')

Nodo inicial:  ORADEA
Nodo objetivo:  FAGARAS


Ruta en la cola:  ['ORADEA']  - Último nodo de la ruta  ORADEA
Nodo no explorado ORADEA 
Nodos vecinos de :  ORADEA  son:  {'ZERIND', 'SIBIU'}
Nuevas Rutas ['ORADEA', 'ZERIND']
Nuevas Rutas ['ORADEA', 'SIBIU']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['ORADEA', 'ZERIND'], ['ORADEA', 'SIBIU']]
-------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['ORADEA', 'ZERIND']  - Último nodo de la ruta  ZERIND
Nodo no explorado ZERIND 
Nodos vecinos de :  ZERIND  son:  {'ORADEA', 'ARAD'}
Nuevas Rutas ['ORADEA', 'ZERIND', 'ORADEA']
Nuevas Rutas ['ORADEA', 'ZERIND', 'ARAD']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['ORADEA', 'SIBIU'], ['ORADEA', 'ZERIND', 'ORADEA'], ['ORADEA', 'ZERIND', 'ARAD']]
-------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['ORADEA', 'SIBIU']  -

('Camino Escogido: ', ['ORADEA', 'SIBIU', 'FAGARAS'])

**PRUEBA 9 BFS**

En la prueba 9 se puede observar que el nodo de inicio es RIMNICU VILCEA recorre el grafo en anchura y encuentra el camino más corto tal que queue va registrando las rutas posibles en lugar de nodos. Cuando el algoritmo verifica si hay un nodo vecino, debe verificar si el nodo vecino corresponde al nodo objetivo. Esta prueba el objetivo de MEHADIA si se encuentra ya no hay necesidad de seguir explorando el gráfo.

In [0]:
bfs_shortest_path(graph,'RIMNICU VILCEA','MEHADIA')

Nodo inicial:  RIMNICU VILCEA
Nodo objetivo:  MEHADIA


Ruta en la cola:  ['RIMNICU VILCEA']  - Último nodo de la ruta  RIMNICU VILCEA
Nodo no explorado RIMNICU VILCEA 
Nodos vecinos de :  RIMNICU VILCEA  son:  {'SIBIU', 'CRAIOVA', 'PITESTI'}
Nuevas Rutas ['RIMNICU VILCEA', 'SIBIU']
Nuevas Rutas ['RIMNICU VILCEA', 'CRAIOVA']
Nuevas Rutas ['RIMNICU VILCEA', 'PITESTI']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['RIMNICU VILCEA', 'SIBIU'], ['RIMNICU VILCEA', 'CRAIOVA'], ['RIMNICU VILCEA', 'PITESTI']]
-------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['RIMNICU VILCEA', 'SIBIU']  - Último nodo de la ruta  SIBIU
Nodo no explorado SIBIU 
Nodos vecinos de :  SIBIU  son:  {'RIMNICU VILCEA', 'ORADEA', 'ARAD', 'FAGARAS'}
Nuevas Rutas ['RIMNICU VILCEA', 'SIBIU', 'RIMNICU VILCEA']
Nuevas Rutas ['RIMNICU VILCEA', 'SIBIU', 'ORADEA']
Nuevas Rutas ['RIMNICU VILCEA', 'SIBIU', 'ARAD']
Nuevas Rutas ['RIMNICU VIL

('Camino Escogido: ', ['RIMNICU VILCEA', 'CRAIOVA', 'DOBRETA', 'MEHADIA'])

**PRUEBA 10 BFS**

En la prueba 10 se puede observar que el nodo de inicio es SIBIU recorre el grafo en anchura y encuentra el camino más corto tal que queue va registrando las rutas posibles en lugar de nodos. Cuando el algoritmo verifica si hay un nodo vecino, debe verificar si el nodo vecino corresponde al nodo objetivo. Esta prueba el objetivo de GIURGIU si se encuentra ya no hay necesidad de seguir explorando el gráfo.

In [0]:
bfs_shortest_path(graph,'SIBIU','GIURGIU')

Nodo inicial:  SIBIU
Nodo objetivo:  GIURGIU


Ruta en la cola:  ['SIBIU']  - Último nodo de la ruta  SIBIU
Nodo no explorado SIBIU 
Nodos vecinos de :  SIBIU  son:  {'RIMNICU VILCEA', 'ORADEA', 'ARAD', 'FAGARAS'}
Nuevas Rutas ['SIBIU', 'RIMNICU VILCEA']
Nuevas Rutas ['SIBIU', 'ORADEA']
Nuevas Rutas ['SIBIU', 'ARAD']
Nuevas Rutas ['SIBIU', 'FAGARAS']
No se encontro Objetivo
Nuevas rutas agregada a la cola:  [['SIBIU', 'RIMNICU VILCEA'], ['SIBIU', 'ORADEA'], ['SIBIU', 'ARAD'], ['SIBIU', 'FAGARAS']]
-------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['SIBIU', 'RIMNICU VILCEA']  - Último nodo de la ruta  RIMNICU VILCEA
Nodo no explorado RIMNICU VILCEA 
Nodos vecinos de :  RIMNICU VILCEA  son:  {'SIBIU', 'CRAIOVA', 'PITESTI'}
Nuevas Rutas ['SIBIU', 'RIMNICU VILCEA', 'SIBIU']
Nuevas Rutas ['SIBIU', 'RIMNICU VILCEA', 'CRAIOVA']
Nuevas Rutas ['SIBIU', 'RIMNICU VILCEA', 'PITESTI']
No se encontro Objetivo
Nuevas rutas 

('Camino Escogido: ', ['SIBIU', 'FAGARAS', 'BUCHAREST', 'GIURGIU'])

# ALGORITMO DSF

Se realizó la función dfs la cual implementa una búsqueda en profundidad. Su funcionamiento consiste en ir expandiendo todos y cada uno de los nodos que va localizando, de forma recurrente, en un camino concreto. Cuando ya no quedan más nodos que visitar en dicho camino, regresa de modo que repite el mismo proceso con cada uno de los nodos vecinos ya procesado.

**PRUEBA 1 DFS**

In [0]:
def dfs_paths(graph, start, goal):
  
    print('Nodo inicial: ',start)
    print ('Nodo objetivo: ',goal)
    print('\n')
  
    # Define stack variable
    stack = [[start]]
    # Do the process while there are paths to follow
    while stack:
        path = stack.pop()
        node = path[-1]
        
        print('Ruta en la cola: ',path, 'Último nodo en la ruta: ',node)
        for next in graph[node] - set(path):
            # If a correct path is founded, then return the path with the generator
            # else write a new path and follow iterating.
            if next == goal:
                print('Camino Escogido: ', path + [next], ' - Objetivo Encontrado: ',next)
                print('------------------------------------------------------------------------------------------------------------------------')
                path + [next]
            else:
                stack.append(path + [next])

En la prueba 1 se puede visualizar que el nodo de inicio es ORADEA este a diferencia del BFS recorre el grafo en profundidad va expandiendo todos y cada uno de los nodos que va localizando de forma repetitivo en un camino en concreto. Si ya no quedan mas nodos que visitar por ese camino, se regresa al anterior y este vuelve a repite el mismo proceso con todos los nodos vecinos ya explorado.


In [0]:
  dfs_paths(graph,'ORADEA','SIBIU')

Nodo inicial:  ORADEA
Nodo objetivo:  SIBIU


Ruta en la cola:  ['ORADEA'] Último nodo en la ruta:  ORADEA
Camino Escogido:  ['ORADEA', 'SIBIU']  - Objetivo Encontrado:  SIBIU
------------------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['ORADEA', 'ZERIND'] Último nodo en la ruta:  ZERIND
Ruta en la cola:  ['ORADEA', 'ZERIND', 'ARAD'] Último nodo en la ruta:  ARAD
Camino Escogido:  ['ORADEA', 'ZERIND', 'ARAD', 'SIBIU']  - Objetivo Encontrado:  SIBIU
------------------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['ORADEA', 'ZERIND', 'ARAD', 'TIMISOARA'] Último nodo en la ruta:  TIMISOARA
Ruta en la cola:  ['ORADEA', 'ZERIND', 'ARAD', 'TIMISOARA', 'LUGOJ'] Último nodo en la ruta:  LUGOJ
Ruta en la cola:  ['ORADEA', 'ZERIND', 'ARAD', 'TIMISOARA', 'LUGOJ', 'MEHADIA'] Último nodo en la ruta:  MEHADIA
Ruta en la cola:  ['ORADEA', 'ZERIND', 

**PRUEBA 2 DFS**

En la prueba 2 se puede visualizar que el nodo de inicio es HIRSOVA este a diferencia del BFS recorre el grafo en profundidad va expandiendo todos y cada uno de los nodos que va localizando de forma repetitivo en un camino en concreto. Si ya no quedan mas nodos que visitar por ese camino, se regresa al anterior y este vuelve a repite el mismo proceso con todos los nodos vecinos ya explorado.


In [0]:
  dfs_paths(graph,'HIRSOVA','URZICENI')

Nodo inicial:  HIRSOVA
Nodo objetivo:  URZICENI


Ruta en la cola:  ['HIRSOVA'] Último nodo en la ruta:  HIRSOVA
Camino Escogido:  ['HIRSOVA', 'URZICENI']  - Objetivo Encontrado:  URZICENI
------------------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['HIRSOVA', 'EFORIE'] Último nodo en la ruta:  EFORIE


**PRUEBA 3 DFS**

En la prueba 3 se puede visualizar que el nodo de inicio es EFORIE este a diferencia del BFS recorre el grafo en profundidad va expandiendo todos y cada uno de los nodos que va localizando de forma repetitivo en un camino en concreto. Si ya no quedan mas nodos que visitar por ese camino, se regresa al anterior y este vuelve a repite el mismo proceso con todos los nodos vecinos ya explorado.


In [0]:
  dfs_paths(graph,'EFORIE','SIBIU')

Nodo inicial:  EFORIE
Nodo objetivo:  SIBIU


Ruta en la cola:  ['EFORIE'] Último nodo en la ruta:  EFORIE
Ruta en la cola:  ['EFORIE', 'HIRSOVA'] Último nodo en la ruta:  HIRSOVA
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI'] Último nodo en la ruta:  URZICENI
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'VASLUI'] Último nodo en la ruta:  VASLUI
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'VASLUI', 'IASI'] Último nodo en la ruta:  IASI
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'VASLUI', 'IASI', 'NEAMT'] Último nodo en la ruta:  NEAMT
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'BUCHAREST'] Último nodo en la ruta:  BUCHAREST
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'BUCHAREST', 'PITESTI'] Último nodo en la ruta:  PITESTI
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'BUCHAREST', 'PITESTI', 'CRAIOVA'] Último nodo en la ruta:  CRAIOVA
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'BUCHAREST', 'PITESTI', 'CRAIOVA', 'RIMNICU VILCE

**PRUEBA 4 DFS**

En la prueba 4 se puede visualizar que el nodo de inicio es DOBRETA este a diferencia del BFS recorre el grafo en profundidad va expandiendo todos y cada uno de los nodos que va localizando de forma repetitivo en un camino en concreto. Si ya no quedan mas nodos que visitar por ese camino, se regresa al anterior y este vuelve a repite el mismo proceso con todos los nodos vecinos ya explorado.


In [0]:
  dfs_paths(graph,'DOBRETA','BUCHAREST')

Nodo inicial:  DOBRETA
Nodo objetivo:  BUCHAREST


Ruta en la cola:  ['DOBRETA'] Último nodo en la ruta:  DOBRETA
Ruta en la cola:  ['DOBRETA', 'CRAIOVA'] Último nodo en la ruta:  CRAIOVA
Ruta en la cola:  ['DOBRETA', 'CRAIOVA', 'PITESTI'] Último nodo en la ruta:  PITESTI
Camino Escogido:  ['DOBRETA', 'CRAIOVA', 'PITESTI', 'BUCHAREST']  - Objetivo Encontrado:  BUCHAREST
------------------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['DOBRETA', 'CRAIOVA', 'PITESTI', 'RIMNICU VILCEA'] Último nodo en la ruta:  RIMNICU VILCEA
Ruta en la cola:  ['DOBRETA', 'CRAIOVA', 'PITESTI', 'RIMNICU VILCEA', 'SIBIU'] Último nodo en la ruta:  SIBIU
Ruta en la cola:  ['DOBRETA', 'CRAIOVA', 'PITESTI', 'RIMNICU VILCEA', 'SIBIU', 'FAGARAS'] Último nodo en la ruta:  FAGARAS
Camino Escogido:  ['DOBRETA', 'CRAIOVA', 'PITESTI', 'RIMNICU VILCEA', 'SIBIU', 'FAGARAS', 'BUCHAREST']  - Objetivo Encontrado:  BUCHAREST
-------------------------

**PRUEBA 5 DFS**

En la prueba 5 se puede visualizar que el nodo de inicio es DOBRETA este a diferencia del BFS recorre el grafo en profundidad va expandiendo todos y cada uno de los nodos que va localizando de forma repetitivo en un camino en concreto. Si ya no quedan mas nodos que visitar por ese camino, se regresa al anterior y este vuelve a repite el mismo proceso con todos los nodos vecinos ya explorado.


In [0]:
  dfs_paths(graph,'DOBRETA','PITESTI')

Nodo inicial:  DOBRETA
Nodo objetivo:  PITESTI


Ruta en la cola:  ['DOBRETA'] Último nodo en la ruta:  DOBRETA
Ruta en la cola:  ['DOBRETA', 'CRAIOVA'] Último nodo en la ruta:  CRAIOVA
Camino Escogido:  ['DOBRETA', 'CRAIOVA', 'PITESTI']  - Objetivo Encontrado:  PITESTI
------------------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['DOBRETA', 'CRAIOVA', 'RIMNICU VILCEA'] Último nodo en la ruta:  RIMNICU VILCEA
Camino Escogido:  ['DOBRETA', 'CRAIOVA', 'RIMNICU VILCEA', 'PITESTI']  - Objetivo Encontrado:  PITESTI
------------------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['DOBRETA', 'CRAIOVA', 'RIMNICU VILCEA', 'SIBIU'] Último nodo en la ruta:  SIBIU
Ruta en la cola:  ['DOBRETA', 'CRAIOVA', 'RIMNICU VILCEA', 'SIBIU', 'FAGARAS'] Último nodo en la ruta:  FAGARAS
Ruta en la cola:  ['DOBRETA', 'CRAIOVA', 'RIMNICU VILCEA', 'SIBIU', 'FAGAR

**PRUEBA 6 DFS**

En la prueba 6 se puede visualizar que el nodo de inicio es EFORIE este a diferencia del BFS recorre el grafo en profundidad va expandiendo todos y cada uno de los nodos que va localizando de forma repetitivo en un camino en concreto. Si ya no quedan mas nodos que visitar por ese camino, se regresa al anterior y este vuelve a repite el mismo proceso con todos los nodos vecinos ya explorado.


In [0]:
  dfs_paths(graph,'EFORIE','BUCHAREST')

Nodo inicial:  EFORIE
Nodo objetivo:  BUCHAREST


Ruta en la cola:  ['EFORIE'] Último nodo en la ruta:  EFORIE
Ruta en la cola:  ['EFORIE', 'HIRSOVA'] Último nodo en la ruta:  HIRSOVA
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI'] Último nodo en la ruta:  URZICENI
Camino Escogido:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'BUCHAREST']  - Objetivo Encontrado:  BUCHAREST
------------------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'VASLUI'] Último nodo en la ruta:  VASLUI
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'VASLUI', 'IASI'] Último nodo en la ruta:  IASI
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'VASLUI', 'IASI', 'NEAMT'] Último nodo en la ruta:  NEAMT


**PRUEBA 7 DFS**

En la prueba 7 se puede visualizar que el nodo de inicio es BUCHAREST este a diferencia del BFS recorre el grafo en profundidad va expandiendo todos y cada uno de los nodos que va localizando de forma repetitivo en un camino en concreto. Si ya no quedan mas nodos que visitar por ese camino, se regresa al anterior y este vuelve a repite el mismo proceso con todos los nodos vecinos ya explorado.


In [0]:
  dfs_paths(graph,'BUCHAREST','LUGOJ')

Nodo inicial:  BUCHAREST
Nodo objetivo:  LUGOJ


Ruta en la cola:  ['BUCHAREST'] Último nodo en la ruta:  BUCHAREST
Ruta en la cola:  ['BUCHAREST', 'PITESTI'] Último nodo en la ruta:  PITESTI
Ruta en la cola:  ['BUCHAREST', 'PITESTI', 'CRAIOVA'] Último nodo en la ruta:  CRAIOVA
Ruta en la cola:  ['BUCHAREST', 'PITESTI', 'CRAIOVA', 'RIMNICU VILCEA'] Último nodo en la ruta:  RIMNICU VILCEA
Ruta en la cola:  ['BUCHAREST', 'PITESTI', 'CRAIOVA', 'RIMNICU VILCEA', 'SIBIU'] Último nodo en la ruta:  SIBIU
Ruta en la cola:  ['BUCHAREST', 'PITESTI', 'CRAIOVA', 'RIMNICU VILCEA', 'SIBIU', 'FAGARAS'] Último nodo en la ruta:  FAGARAS
Ruta en la cola:  ['BUCHAREST', 'PITESTI', 'CRAIOVA', 'RIMNICU VILCEA', 'SIBIU', 'ARAD'] Último nodo en la ruta:  ARAD
Ruta en la cola:  ['BUCHAREST', 'PITESTI', 'CRAIOVA', 'RIMNICU VILCEA', 'SIBIU', 'ARAD', 'TIMISOARA'] Último nodo en la ruta:  TIMISOARA
Camino Escogido:  ['BUCHAREST', 'PITESTI', 'CRAIOVA', 'RIMNICU VILCEA', 'SIBIU', 'ARAD', 'TIMISOARA', 'LUGOJ']  - Ob

**PRUEBA 8 DFS**

En la prueba 1 se puede visualizar que el nodo de inicio es GIURGIU este a diferencia del BFS recorre el grafo en profundidad va expandiendo todos y cada uno de los nodos que va localizando de forma repetitivo en un camino en concreto. Si ya no quedan mas nodos que visitar por ese camino, se regresa al anterior y este vuelve a repite el mismo proceso con todos los nodos vecinos ya explorado.


In [0]:
  dfs_paths(graph,'GIURGIU','CRAIOVA')

Nodo inicial:  GIURGIU
Nodo objetivo:  CRAIOVA


Ruta en la cola:  ['GIURGIU'] Último nodo en la ruta:  GIURGIU
Ruta en la cola:  ['GIURGIU', 'BUCHAREST'] Último nodo en la ruta:  BUCHAREST
Ruta en la cola:  ['GIURGIU', 'BUCHAREST', 'PITESTI'] Último nodo en la ruta:  PITESTI
Camino Escogido:  ['GIURGIU', 'BUCHAREST', 'PITESTI', 'CRAIOVA']  - Objetivo Encontrado:  CRAIOVA
------------------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['GIURGIU', 'BUCHAREST', 'PITESTI', 'RIMNICU VILCEA'] Último nodo en la ruta:  RIMNICU VILCEA
Camino Escogido:  ['GIURGIU', 'BUCHAREST', 'PITESTI', 'RIMNICU VILCEA', 'CRAIOVA']  - Objetivo Encontrado:  CRAIOVA
------------------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['GIURGIU', 'BUCHAREST', 'PITESTI', 'RIMNICU VILCEA', 'SIBIU'] Último nodo en la ruta:  SIBIU
Ruta en la cola:  ['GIURGIU', 'BUCHAREST', 

**PRUEBA 9 DFS**

En la prueba 9 se puede visualizar que el nodo de inicio es URZICENI este a diferencia del BFS recorre el grafo en profundidad va expandiendo todos y cada uno de los nodos que va localizando de forma repetitivo en un camino en concreto. Si ya no quedan mas nodos que visitar por ese camino, se regresa al anterior y este vuelve a repite el mismo proceso con todos los nodos vecinos ya explorado.


In [0]:
  dfs_paths(graph,'URZICENI','PITESTI')

Nodo inicial:  URZICENI
Nodo objetivo:  PITESTI


Ruta en la cola:  ['URZICENI'] Último nodo en la ruta:  URZICENI
Ruta en la cola:  ['URZICENI', 'VASLUI'] Último nodo en la ruta:  VASLUI
Ruta en la cola:  ['URZICENI', 'VASLUI', 'IASI'] Último nodo en la ruta:  IASI
Ruta en la cola:  ['URZICENI', 'VASLUI', 'IASI', 'NEAMT'] Último nodo en la ruta:  NEAMT
Ruta en la cola:  ['URZICENI', 'HIRSOVA'] Último nodo en la ruta:  HIRSOVA
Ruta en la cola:  ['URZICENI', 'HIRSOVA', 'EFORIE'] Último nodo en la ruta:  EFORIE
Ruta en la cola:  ['URZICENI', 'BUCHAREST'] Último nodo en la ruta:  BUCHAREST
Camino Escogido:  ['URZICENI', 'BUCHAREST', 'PITESTI']  - Objetivo Encontrado:  PITESTI
------------------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['URZICENI', 'BUCHAREST', 'FAGARAS'] Último nodo en la ruta:  FAGARAS
Ruta en la cola:  ['URZICENI', 'BUCHAREST', 'FAGARAS', 'SIBIU'] Último nodo en la ruta:  SIBIU
Ruta en la col

**PRUEBA 10 DFS**

En la prueba 10 se puede visualizar que el nodo de inicio es EFORIE este a diferencia del BFS recorre el grafo en profundidad va expandiendo todos y cada uno de los nodos que va localizando de forma repetitivo en un camino en concreto. Si ya no quedan mas nodos que visitar por ese camino, se regresa al anterior y este vuelve a repite el mismo proceso con todos los nodos vecinos ya explorado.


In [0]:
  dfs_paths(graph,'EFORIE','VASLUI')

Nodo inicial:  EFORIE
Nodo objetivo:  VASLUI


Ruta en la cola:  ['EFORIE'] Último nodo en la ruta:  EFORIE
Ruta en la cola:  ['EFORIE', 'HIRSOVA'] Último nodo en la ruta:  HIRSOVA
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI'] Último nodo en la ruta:  URZICENI
Camino Escogido:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'VASLUI']  - Objetivo Encontrado:  VASLUI
------------------------------------------------------------------------------------------------------------------------
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'BUCHAREST'] Último nodo en la ruta:  BUCHAREST
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'BUCHAREST', 'PITESTI'] Último nodo en la ruta:  PITESTI
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'BUCHAREST', 'PITESTI', 'CRAIOVA'] Último nodo en la ruta:  CRAIOVA
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZICENI', 'BUCHAREST', 'PITESTI', 'CRAIOVA', 'RIMNICU VILCEA'] Último nodo en la ruta:  RIMNICU VILCEA
Ruta en la cola:  ['EFORIE', 'HIRSOVA', 'URZ

# **ALGORITMO DE BUSQUEDA A***

Primero se realizará la representación de un problema de búsqueda, la cual un problema de búsqueda consiste en: 
 	un nodo de inicio (start):
  
*   una función de vecinos (neighbors) que da los vecinos (neighbors) de un nodo
*   una especificación de un objetivo
*   una función heurística (opcional).

Los métodos deben ser sobrescritos para definir un problema de búsqueda:

In [0]:
class Search_problem(object):
    """Un problema de búsqueda consiste de:
    * un nodo de inicio (start)
    * una funcioón de vecinos (neighbors) que da los vecinos (neighbors) de un nodo
    * una especificación de un objetivo
    * una función heurística (opcional).
    Los métodos deben ser sobrescritos para definir un problema de búsqueda."""

    def start_node(self):
        """retorna el node inicial (start)"""
        raise NotImplementedError("start_node")   # abstract method
    
    def is_goal(self,node):
        """es True si nodo es un objetivo"""
        raise NotImplementedError("is_goal")   # abstract method

    def neighbors(self,node):
        """retorna una lista de los  arcos para los vecinos (neighbors) de un nodo"""
        raise NotImplementedError("neighbors")   # abstract method

    def heuristic(self,n):
        """Da el valor heuristico del nodo n.
        Retorna 0 si no se sobrescribe."""
        return 0


Luego tenemos un arco que tiene un nodo from_node y un nodo to_node y un costo (no negativo):

In [0]:
class Arc(object):
    """Un arco tiene un nodo from_node y un nodo to_node y un costo (no-negativo)"""
    def __init__(self, from_node, to_node, cost=1, action=None):
        assert cost >= 0, ("Costo no puede ser negativo para"+
                           str(from_node)+"->"+str(to_node)+", cost: "+str(cost))
        self.from_node = from_node
        self.to_node = to_node
        self.action = action
        self.cost=cost

    def __repr__(self):
        """Retorna un  string de la representación de un arco"""
        if self.action:
            return str(self.from_node)+" --"+str(self.action)+"--> "+str(self.to_node)
        else:
            return str(self.from_node)+" --> "+str(self.to_node)
            


En la clase de representación explicita del grafo de búsqueda un problema de búsqueda consiste de:
*  una lista o set de nodos: nodes
*	una lista o set de arcos: arcs
*	un nodo inicial (start):  start=None
*	una lista o set de nodos objetivos: goals=set()
* un diccionario que mapea cada nodo en su valor heurístico: hmap={}

Se define un problema de búsqueda, necesitamos definir el nodo de inicio, el predicado objetivo, la función de los vecinos y la función heurística. Lo siguiente se usa para la implementación de búsqueda (DFS) primero en profundidad a continuación 

In [0]:
class Search_problem_from_explicit_graph(Search_problem):
    """Un problema de búsqueda consiste de:
    * una lista o set de nodos: nodes
    * una lista o set de arcos: arcs
    * un nodo inicial (start):  start=None
    * una lista o set de nodos objetivos: goals=set()
    * un diccionario que mapea cada nodo en su valor heurístico: hmap={}
    """
  

    def __init__(self, nodes, arcs, start=None, goals=set(), hmap={}):
        self.neighs = {}
        self.nodes = nodes
               

        for node in nodes:
            self.neighs[node]=[]
        self.arcs = arcs
        for arc in arcs:
            self.neighs[arc.from_node].append(arc)
        self.start = start
        self.goals = goals
        self.hmap = hmap

    def start_node(self):
        """retorna el nodo inicial (start)"""
        return self.start
    
    def is_goal(self,node):
        """es True so nodo es un objetivo (goal)"""
        return node in self.goals

    def neighbors(self,node):
        """retorna los vecinos (neighbors) de un nodo"""
        return self.neighs[node]

    def heuristic(self,node):
        """Dado el valor heurístico de un nodo n.
        Retorna 0 si no sobrescribir en el hmap."""
        if node in self.hmap:
            return self.hmap[node]
        else:
            return 0
        
    def __repr__(self):
        """retorna un string de la representación del problema de búsqueda"""
        res=""
        for arc in self.arcs:
            res += str(arc)+".  "
        return res

    def neighbor_nodes(self,node):
        """retorna un iterador sobre los vecinos (neighbors) de un nodo"""
        return (path.to_node for path in self.neighs[node])



La cual un buscador devolvera una ruta desde el nodo de inicio a un nodo objetivo. Se utiliza una representacion alternativa en terminos de una estructura de datos recursiva que puede compartir subpartes. Un camino es:
* un nodo (que representa una ruta de longitud 0) o
*	una ruta, inicial y un arco, donde el nodo desde el arco es el nodo en el final de la inicial.

Estos casos se distinguen en el siguiente código al tener arc = None si la ruta tiene una longitud de 0, en cuyo caso el nodo inicial (start) es el nodo de la ruta. 


In [0]:
class Path(object):
    """Una ruta (path) es ya sea un nodo o una trayectoria seguida por un arco"""
    
    def __init__(self,initial,arc=None):
        """inicial es un nodo (en cuyo caso el arco es None) o
        una ruta (path) (en cuyo caso arco es un objeto de tipo arco)"""
        self.initial = initial
        self.arc=arc
        if arc is None:
            self.cost=0
        else:
            self.cost = initial.cost + arc.cost

    def end(self):
        """retorna el nodo en el fin de la ruta (path)"""
        if self.arc is None:
            return self.initial
        else:
            return self.arc.to_node

    def nodes(self):
        """enumera los nodos para la ruta.
        Esto comienza al final y enumera los nodos en la ruta hacia atrás."""
        current = self
        while current.arc is not None:
            yield current.arc.to_node
            current = current.initial
        yield current.initial

    def initial_nodes(self):
        """enumera los nodos para la ruta antes del nodo final.
        Esto comienza al final y enumera los nodos en la ruta hacia atrás."""
        if self.arc is not None:
            for nd in self.initial.nodes(): yield nd     # could be "yield from"
        
    def __repr__(self):
        """retorna un string de representación de una ruta (path)"""
        if self.arc is None:
            return str(self.initial)
        elif self.arc.action:
            return (str(self.initial)+"\n   --"+str(self.arc.action)
                    +"--> "+str(self.arc.to_node))
        else:
            return str(self.initial)+" --> "+str(self.arc.to_node)



Se construye un objeto Searcher para el problema y luego pedir repetidamente la siguiente ruta utilizando la búsqueda. Si no hay más caminos, se devuelve None

In [0]:
class Displayable(object):
    """Class that uses 'display'.
    The amount of detail is controlled by max_display_level
    """
    max_display_level = 1   # can be overridden in subclasses

    def display(self,level, *args, **nargs):
        """print the arguments if level is less than or equal to the
        current max_display_level.
        level is an integer.
        the other arguments are whatever arguments print can take.
        """
        if level <= self.max_display_level:
            print(*args, **nargs)  ##if error you are using Python2 not Python3
def visualize(func):
    """A decorator for algorithms that do interactive visualization.
    Ignored here.
    """
    return func

In [0]:
class Searcher(Displayable):
    """retorna un buscador (searcher) para un problema.
    Las Rutas (Paths) pueden ser encontradas por llamadas repetidas a search().
    Esto hace primero la búsqueda de profundidad a menos que se anule
    """
    def __init__(self, problem):
        """crea un buscador (searcher) desde un problema
        """
        self.problem = problem
        self.initialize_frontier()
        self.num_expanded = 0
        self.add_to_frontier(Path(problem.start_node()))
        super().__init__()

    def initialize_frontier(self):
        self.frontier = []
        
    def empty_frontier(self):
        return self.frontier == []
        
    def add_to_frontier(self,path):
        self.frontier.append(path)
        
    @visualize
    def search(self):
        """retorna (next) path - (siguiente) ruta - desde el nodo inicial (start) de un problema
        a un nodo objetivo (goal). 
        Retorna None si no existe ruta (path).
        """
        while not self.empty_frontier():
          
            path = self.frontier.pop()
            
            self.display(2, "Expandiendo:",path,"(costo:",path.cost,")")
            self.num_expanded += 1
            
            if self.problem.is_goal(path.end()):    # solución encontrada
                self.display(1, self.num_expanded, "caminos se han ampliado y",
                            len(self.frontier), "rutas (paths) permanecen en la frontera")
                self.solution = path   # almacena la solución encontrada
                return path
            else:
                neighs = self.problem.neighbors(path.end())
                self.display(3,"Los vecinos (Neighbors) son", neighs)
                
                for arc in reversed(neighs):
                    self.add_to_frontier(Path(path,arc))
                self.display(3,"Frontera:",self.frontier)
                
        self.display(1,"No (más) soluciones. Total de",
                     self.num_expanded,"rutas (paths) expandidas.")


En la clase frontera consiste de una cola (queue) de prioridas (priority) (heap), fronterapq, de triples (value, index, path), dónde:
* 	value es el valor que queremos minimizar (e.g., costo de path + h).
*  index es un índice único para cada elemento
* path es la ruta en la cola (queue)

Tenga en cuenta que la cola de prioridad siempre devuelve el elemento más pequeño

In [0]:
import heapq        

#from searchProblem import Path

class FrontierPQ(object):
    """Una frontera consiste de una cola (queue) de prioridas (priority) (heap), fronterapq, de
       triples (value, index, path), dónde
    * value es el valor que queremos minimizar (e.g., costo de path + h).
    * index es un índice único para cada elemento
    * path es la ruta en la cola (queue)
    Tenga en cuenta que la cola de prioridad siempre devuelve el elemento más pequeño.
    """

    def __init__(self):
        """construye la frontera, inicialmente una cola de prioridad vacía
        """
        self.frontier_index = 0  # el número de elementos que se han añadido a la frontera.
        self.frontierpq = []  # la cola de prioridad de la frontera

    def empty(self):
        """es True si la cola de priority está vacía (empty)"""
        return self.frontierpq == []

    def add(self, path, value):
        """añade una ruta (path) a la cola priority
        value es el valor a ser minimizado"""
        self.frontier_index += 1    # obtener un nuevo índice único
        heapq.heappush(self.frontierpq,(value, -self.frontier_index, path))

    def pop(self):
        """retorna y elimina el camino de la frontera con valor mínimo.
        """
        (_,_,path) = heapq.heappop(self.frontierpq)
        return path 

    def count(self,val):
        """retorna el número de elementos de la frontera con value=val"""
        return sum(1 for e in self.frontierpq if e[0]==val)

    def __repr__(self):
        """string de representación de la frontier"""
        return str([(n,c,str(p)) for (n,c,p) in self.frontierpq])
    
    def __len__(self):
        """longitud (length) de la frontier"""
        return len(self.frontierpq)

    def __iter__(self):
        """iterar a través de a través de los caminos en el frontier"""
        for (_,_,path) in self.frontierpq:
            yield path

Para una búsqueda A∗, la frontera se implementa utilizando la clase FrontierPQ. 

In [0]:
class AStarSearcher(Searcher):
    """retorna un buscador searcher para un problema.
    Las rutas se pueden encontrar llamando repetidamente a search().
    """

    def __init__(self, problem):
        super().__init__(problem)

    def initialize_frontier(self):
        self.frontier = FrontierPQ()

    def empty_frontier(self):
        return self.frontier.empty()

    def add_to_frontier(self,path):
        """agregar ruta (path) a la frontera con el costo apropiado"""
        value = path.cost + self.problem.heuristic(path.end())
        print('Costo del camino',path.cost,' + Costo de la heurística de: ',path.end(),' es: ',self.problem.heuristic(path.end()),' --> ',value)
        self.frontier.add(path, value)

El problema de entrega es en forma cíclica descrito en el problemageneral. Por que todos los arcos también tienen su inverso.

In [0]:
problemageneral = Search_problem_from_explicit_graph(
    
   {'ARAD','BUCHAREST','CRAIOVA','DOBRETA','EFOIRE','FAGARAS','GIURGIU','HIRSOVA','IASI' ,'LUGOJ','MEHADIA','NEAMT','ORADEA','PITESTI','RIMNICU VILCEA','SIBIU',
    'TIMISOARA','URZICENI', 'VASLUI','ZERIND'},
     [  Arc('ARAD','ZERIND',75), Arc('ZERIND','ARAD',75),
        Arc('ARAD','SIBIU',140), Arc('SIBIU','ARAD',140),
        Arc('ARAD','TIMISOARA',118),Arc('TIMISOARA','ARAD',118),  
        Arc('BUCHAREST','PITESTI',101), Arc('PITESTI','BUCHAREST',101),
        Arc('BUCHAREST','FAGARAS',211), Arc('FAGARAS','BUCHAREST',211),
        Arc('BUCHAREST','URZICENI',85), Arc('URZICENI','BUCHAREST',85),
        Arc('BUCHAREST','GIURGIU',90),Arc('GIURGIU','BUCHAREST',90),
        Arc('CRAIOVA','PITESTI',138), Arc('PITESTI','CRAIOVA',138),
        Arc('CRAIOVA','RIMNICU VILCEA',146), Arc('RIMNICU VILCEA','CRAIOVA',146),
        Arc('CRAIOVA','DOBRETA',120), Arc('DOBRETA','CRAIOVA',120),
        Arc('DOBRETA','MEHADIA',75), Arc('MEHADIA','DOBRETA',75),
        Arc('EFOIRE','HIRSOVA',86),  Arc('HIRSOVA','EFOIRE',86),
        Arc('FAGARAS','SIBIU',99), Arc('SIBIU','FAGARAS',99),
        Arc('HIRSOVA','URZICENI',98), Arc('URZICENI','HIRSOVA',98),
        Arc('IASI','VASLUI',92), Arc('VASLUI','IASI',92),
        Arc('IASI','NEAMT',87), Arc('NEAMT','IASI',87),
        Arc('LUGOJ','MEHADIA',70), Arc('MEHADIA','LUGOJ',70),
        Arc('LUGOJ','TIMISOARA',111), Arc('TIMISOARA','LUGOJ',111),
        Arc('ORADEA','ZERIND',71), Arc('ZERIND','ORADEA',71),
        Arc('ORADEA','SIBIU',151), Arc('SIBIU','ORADEA',151),
        Arc('PITESTI','RIMNICU VILCEA',97), Arc('RIMNICU VILCEA','PITESTI',97),
        Arc('RIMNICU VILCEA','SIBIU',80), Arc('SIBIU','RIMNICU VILCEA',80),   
        Arc('URZICENI','VASLUI',142), Arc('VASLUI','URZICENI',142)],
    start = 'MEHADIA',
    goals = {'BUCHAREST'},
    hmap = {
        'ARAD' : 366,
        'BUCHAREST' : 0,
        'CRAIOVA' : 160,
        'DOBRETA' : 242,
        'EFOIRE' : 161,
        'FAGARAS' : 178,
        'GIURGIU' : 77,
        'HIRSOVA' : 151,
        'IASI' : 226,
        'LUGOJ' : 244,
        'MEHADIA' : 241,
        'NEAMT' : 234,
        'ORADEA' : 380,
        'PITESTI' : 98,
        'RIMNICU VILCEA' :193,
        'SIBIU' : 253,
        'TIMISOARA' : 329,
        'URZICENI' : 80, 
        'VASLUI' : 199,
        'ZERIND' : 374
        }
    )


**PRUEBA 1 A***

En esta prueba 1 se desea encontrar una solución: secuencia de ciudades que nos lleve desde Arad hasta Bucharest la cual nos dará un costo de la ruta 418


In [0]:
problema1 = Search_problem_from_explicit_graph(
    
   {'ARAD','BUCHAREST','CRAIOVA','DOBRETA','EFOIRE','FAGARAS','GIURGIU','HIRSOVA','IASI' ,'LUGOJ','MEHADIA','NEAMT','ORADEA','PITESTI','RIMNICU VILCEA','SIBIU',
    'TIMISOARA','URZICENI', 'VASLUI','ZERIND'},
     [  Arc('ARAD','ZERIND',75), Arc('ZERIND','ARAD',75),
        Arc('ARAD','SIBIU',140), Arc('SIBIU','ARAD',140),
        Arc('ARAD','TIMISOARA',118),Arc('TIMISOARA','ARAD',118),  
        Arc('BUCHAREST','PITESTI',101), Arc('PITESTI','BUCHAREST',101),
        Arc('BUCHAREST','FAGARAS',211), Arc('FAGARAS','BUCHAREST',211),
        Arc('BUCHAREST','URZICENI',85), Arc('URZICENI','BUCHAREST',85),
        Arc('BUCHAREST','GIURGIU',90),Arc('GIURGIU','BUCHAREST',90),
        Arc('CRAIOVA','PITESTI',138), Arc('PITESTI','CRAIOVA',138),
        Arc('CRAIOVA','RIMNICU VILCEA',146), Arc('RIMNICU VILCEA','CRAIOVA',146),
        Arc('CRAIOVA','DOBRETA',120), Arc('DOBRETA','CRAIOVA',120),
        Arc('DOBRETA','MEHADIA',75), Arc('MEHADIA','DOBRETA',75),
        Arc('EFOIRE','HIRSOVA',86),  Arc('HIRSOVA','EFOIRE',86),
        Arc('FAGARAS','SIBIU',99), Arc('SIBIU','FAGARAS',99),
        Arc('HIRSOVA','URZICENI',98), Arc('URZICENI','HIRSOVA',98),
        Arc('IASI','VASLUI',92), Arc('VASLUI','IASI',92),
        Arc('IASI','NEAMT',87), Arc('NEAMT','IASI',87),
        Arc('LUGOJ','MEHADIA',70), Arc('MEHADIA','LUGOJ',70),
        Arc('LUGOJ','TIMISOARA',111), Arc('TIMISOARA','LUGOJ',111),
        Arc('ORADEA','ZERIND',71), Arc('ZERIND','ORADEA',71),
        Arc('ORADEA','SIBIU',151), Arc('SIBIU','ORADEA',151),
        Arc('PITESTI','RIMNICU VILCEA',97), Arc('RIMNICU VILCEA','PITESTI',97),
        Arc('RIMNICU VILCEA','SIBIU',80), Arc('SIBIU','RIMNICU VILCEA',80),   
        Arc('URZICENI','VASLUI',142), Arc('VASLUI','URZICENI',142)],
    start = 'ARAD',
    goals = {'BUCHAREST'},
    hmap = {
        'ARAD' : 366,
        'BUCHAREST' : 0,
        'CRAIOVA' : 160,
        'DOBRETA' : 242,
        'EFOIRE' : 161,
        'FAGARAS' : 178,
        'GIURGIU' : 77,
        'HIRSOVA' : 151,
        'IASI' : 226,
        'LUGOJ' : 244,
        'MEHADIA' : 241,
        'NEAMT' : 234,
        'ORADEA' : 380,
        'PITESTI' : 98,
        'RIMNICU VILCEA' :193,
        'SIBIU' : 253,
        'TIMISOARA' : 329,
        'URZICENI' : 80, 
        'VASLUI' : 199,
        'ZERIND' : 374
        }
    )


In [0]:
print("Prueba 1: Origen ARAD y destino BUCHAREST")
schr = AStarSearcher(problema1)

path = schr.search()
print("Ruta encontrada:", path )

print('Costo de la Ruta: ', path.cost)

Prueba 1: Origen ARAD y destino BUCHAREST
Costo del camino 0  + Costo de la heurística de:  ARAD  es:  366  -->  366
Costo del camino 118  + Costo de la heurística de:  TIMISOARA  es:  329  -->  447
Costo del camino 140  + Costo de la heurística de:  SIBIU  es:  253  -->  393
Costo del camino 75  + Costo de la heurística de:  ZERIND  es:  374  -->  449
Costo del camino 220  + Costo de la heurística de:  RIMNICU VILCEA  es:  193  -->  413
Costo del camino 291  + Costo de la heurística de:  ORADEA  es:  380  -->  671
Costo del camino 239  + Costo de la heurística de:  FAGARAS  es:  178  -->  417
Costo del camino 280  + Costo de la heurística de:  ARAD  es:  366  -->  646
Costo del camino 300  + Costo de la heurística de:  SIBIU  es:  253  -->  553
Costo del camino 317  + Costo de la heurística de:  PITESTI  es:  98  -->  415
Costo del camino 366  + Costo de la heurística de:  CRAIOVA  es:  160  -->  526
Costo del camino 414  + Costo de la heurística de:  RIMNICU VILCEA  es:  193  -->  60

**PRUEBA 2 A***

En la prueba 2 se desea encontrar una solución: secuencia de ciudades que nos lleve desde Zerind hasta Bucarest.  Para esto se debe cambiar start = ZERIND en el problema de entrega, en esta prueba costo de la ruta de 493 

In [0]:
problema2 = Search_problem_from_explicit_graph(
    
   {'ARAD','BUCHAREST','CRAIOVA','DOBRETA','EFOIRE','FAGARAS','GIURGIU','HIRSOVA','IASI' ,'LUGOJ','MEHADIA','NEAMT','ORADEA','PITESTI','RIMNICU VILCEA','SIBIU',
    'TIMISOARA','URZICENI', 'VASLUI','ZERIND'},
     [  Arc('ARAD','ZERIND',75), Arc('ZERIND','ARAD',75),
        Arc('ARAD','SIBIU',140), Arc('SIBIU','ARAD',140),
        Arc('ARAD','TIMISOARA',118),Arc('TIMISOARA','ARAD',118),  
        Arc('BUCHAREST','PITESTI',101), Arc('PITESTI','BUCHAREST',101),
        Arc('BUCHAREST','FAGARAS',211), Arc('FAGARAS','BUCHAREST',211),
        Arc('BUCHAREST','URZICENI',85), Arc('URZICENI','BUCHAREST',85),
        Arc('BUCHAREST','GIURGIU',90),Arc('GIURGIU','BUCHAREST',90),
        Arc('CRAIOVA','PITESTI',138), Arc('PITESTI','CRAIOVA',138),
        Arc('CRAIOVA','RIMNICU VILCEA',146), Arc('RIMNICU VILCEA','CRAIOVA',146),
        Arc('CRAIOVA','DOBRETA',120), Arc('DOBRETA','CRAIOVA',120),
        Arc('DOBRETA','MEHADIA',75), Arc('MEHADIA','DOBRETA',75),
        Arc('EFOIRE','HIRSOVA',86),  Arc('HIRSOVA','EFOIRE',86),
        Arc('FAGARAS','SIBIU',99), Arc('SIBIU','FAGARAS',99),
        Arc('HIRSOVA','URZICENI',98), Arc('URZICENI','HIRSOVA',98),
        Arc('IASI','VASLUI',92), Arc('VASLUI','IASI',92),
        Arc('IASI','NEAMT',87), Arc('NEAMT','IASI',87),
        Arc('LUGOJ','MEHADIA',70), Arc('MEHADIA','LUGOJ',70),
        Arc('LUGOJ','TIMISOARA',111), Arc('TIMISOARA','LUGOJ',111),
        Arc('ORADEA','ZERIND',71), Arc('ZERIND','ORADEA',71),
        Arc('ORADEA','SIBIU',151), Arc('SIBIU','ORADEA',151),
        Arc('PITESTI','RIMNICU VILCEA',97), Arc('RIMNICU VILCEA','PITESTI',97),
        Arc('RIMNICU VILCEA','SIBIU',80), Arc('SIBIU','RIMNICU VILCEA',80),   
        Arc('URZICENI','VASLUI',142), Arc('VASLUI','URZICENI',142)],
    start = 'ZERIND',
    goals = {'BUCHAREST'},
    hmap = {
        'ARAD' : 366,
        'BUCHAREST' : 0,
        'CRAIOVA' : 160,
        'DOBRETA' : 242,
        'EFOIRE' : 161,
        'FAGARAS' : 178,
        'GIURGIU' : 77,
        'HIRSOVA' : 151,
        'IASI' : 226,
        'LUGOJ' : 244,
        'MEHADIA' : 241,
        'NEAMT' : 234,
        'ORADEA' : 380,
        'PITESTI' : 98,
        'RIMNICU VILCEA' :193,
        'SIBIU' : 253,
        'TIMISOARA' : 329,
        'URZICENI' : 80, 
        'VASLUI' : 199,
        'ZERIND' : 374
        }
    )


In [0]:
print("Prueba 2: Origen ZERIND y destino BUCHAREST")
schr = AStarSearcher(problema2)
path = schr.search()
print("Ruta encontrada:", path)
print('Costo de la Ruta: ', path.cost)

Prueba 2: Origen ZERIND y destino BUCHAREST
Costo del camino 0  + Costo de la heurística de:  TIMISOARA  es:  329  -->  329
Costo del camino 111  + Costo de la heurística de:  LUGOJ  es:  244  -->  355
Costo del camino 118  + Costo de la heurística de:  ARAD  es:  366  -->  484
Costo del camino 222  + Costo de la heurística de:  TIMISOARA  es:  329  -->  551
Costo del camino 181  + Costo de la heurística de:  MEHADIA  es:  241  -->  422
Costo del camino 251  + Costo de la heurística de:  LUGOJ  es:  244  -->  495
Costo del camino 256  + Costo de la heurística de:  DOBRETA  es:  242  -->  498
Costo del camino 236  + Costo de la heurística de:  TIMISOARA  es:  329  -->  565
Costo del camino 258  + Costo de la heurística de:  SIBIU  es:  253  -->  511
Costo del camino 193  + Costo de la heurística de:  ZERIND  es:  374  -->  567
Costo del camino 362  + Costo de la heurística de:  TIMISOARA  es:  329  -->  691
Costo del camino 321  + Costo de la heurística de:  MEHADIA  es:  241  -->  562


**PRUEBA 3 A***

En la prueba 3 se desea encontrar una solución: secuencia de ciudades que nos lleve desde Fagaras hasta Bucarest.  Para esto se debe cambiar start = FAGARAS en el problema de entrega, en esta prueba costo de la ruta de 211 

In [0]:
problema3 = Search_problem_from_explicit_graph(
    
   {'ARAD','BUCHAREST','CRAIOVA','DOBRETA','EFOIRE','FAGARAS','GIURGIU','HIRSOVA','IASI' ,'LUGOJ','MEHADIA','NEAMT','ORADEA','PITESTI','RIMNICU VILCEA','SIBIU',
    'TIMISOARA','URZICENI', 'VASLUI','ZERIND'},
     [  Arc('ARAD','ZERIND',75), Arc('ZERIND','ARAD',75),
        Arc('ARAD','SIBIU',140), Arc('SIBIU','ARAD',140),
        Arc('ARAD','TIMISOARA',118),Arc('TIMISOARA','ARAD',118),  
        Arc('BUCHAREST','PITESTI',101), Arc('PITESTI','BUCHAREST',101),
        Arc('BUCHAREST','FAGARAS',211), Arc('FAGARAS','BUCHAREST',211),
        Arc('BUCHAREST','URZICENI',85), Arc('URZICENI','BUCHAREST',85),
        Arc('BUCHAREST','GIURGIU',90),Arc('GIURGIU','BUCHAREST',90),
        Arc('CRAIOVA','PITESTI',138), Arc('PITESTI','CRAIOVA',138),
        Arc('CRAIOVA','RIMNICU VILCEA',146), Arc('RIMNICU VILCEA','CRAIOVA',146),
        Arc('CRAIOVA','DOBRETA',120), Arc('DOBRETA','CRAIOVA',120),
        Arc('DOBRETA','MEHADIA',75), Arc('MEHADIA','DOBRETA',75),
        Arc('EFOIRE','HIRSOVA',86),  Arc('HIRSOVA','EFOIRE',86),
        Arc('FAGARAS','SIBIU',99), Arc('SIBIU','FAGARAS',99),
        Arc('HIRSOVA','URZICENI',98), Arc('URZICENI','HIRSOVA',98),
        Arc('IASI','VASLUI',92), Arc('VASLUI','IASI',92),
        Arc('IASI','NEAMT',87), Arc('NEAMT','IASI',87),
        Arc('LUGOJ','MEHADIA',70), Arc('MEHADIA','LUGOJ',70),
        Arc('LUGOJ','TIMISOARA',111), Arc('TIMISOARA','LUGOJ',111),
        Arc('ORADEA','ZERIND',71), Arc('ZERIND','ORADEA',71),
        Arc('ORADEA','SIBIU',151), Arc('SIBIU','ORADEA',151),
        Arc('PITESTI','RIMNICU VILCEA',97), Arc('RIMNICU VILCEA','PITESTI',97),
        Arc('RIMNICU VILCEA','SIBIU',80), Arc('SIBIU','RIMNICU VILCEA',80),   
        Arc('URZICENI','VASLUI',142), Arc('VASLUI','URZICENI',142)],
    start = 'FAGARAS',
    goals = {'BUCHAREST'},
    hmap = {
        'ARAD' : 366,
        'BUCHAREST' : 0,
        'CRAIOVA' : 160,
        'DOBRETA' : 242,
        'EFOIRE' : 161,
        'FAGARAS' : 178,
        'GIURGIU' : 77,
        'HIRSOVA' : 151,
        'IASI' : 226,
        'LUGOJ' : 244,
        'MEHADIA' : 241,
        'NEAMT' : 234,
        'ORADEA' : 380,
        'PITESTI' : 98,
        'RIMNICU VILCEA' :193,
        'SIBIU' : 253,
        'TIMISOARA' : 329,
        'URZICENI' : 80, 
        'VASLUI' : 199,
        'ZERIND' : 374
        }
    )


In [0]:
print("Prueba 3: Origen FAGARAS y destino BUCHAREST")
schr = AStarSearcher(problema3)
path = schr.search()
print("Ruta encontrada:", path)
print('Costo de la Ruta: ', path.cost)

Prueba 3: Origen FAGARAS y destino BUCHAREST


NameError: ignored

**PRUEBA 4 A***

En la prueba 4 se desea encontrar una solución: secuencia de ciudades que nos lleve desde Dobreta hasta Bucarest.  Para esto se debe cambiar start = DOBRETA en el problema de entrega, en esta prueba costo de la ruta de 359 

In [0]:
problema4 = Search_problem_from_explicit_graph(
    
   {'ARAD','BUCHAREST','CRAIOVA','DOBRETA','EFOIRE','FAGARAS','GIURGIU','HIRSOVA','IASI' ,'LUGOJ','MEHADIA','NEAMT','ORADEA','PITESTI','RIMNICU VILCEA','SIBIU',
    'TIMISOARA','URZICENI', 'VASLUI','ZERIND'},
     [  Arc('ARAD','ZERIND',75), Arc('ZERIND','ARAD',75),
        Arc('ARAD','SIBIU',140), Arc('SIBIU','ARAD',140),
        Arc('ARAD','TIMISOARA',118),Arc('TIMISOARA','ARAD',118),  
        Arc('BUCHAREST','PITESTI',101), Arc('PITESTI','BUCHAREST',101),
        Arc('BUCHAREST','FAGARAS',211), Arc('FAGARAS','BUCHAREST',211),
        Arc('BUCHAREST','URZICENI',85), Arc('URZICENI','BUCHAREST',85),
        Arc('BUCHAREST','GIURGIU',90),Arc('GIURGIU','BUCHAREST',90),
        Arc('CRAIOVA','PITESTI',138), Arc('PITESTI','CRAIOVA',138),
        Arc('CRAIOVA','RIMNICU VILCEA',146), Arc('RIMNICU VILCEA','CRAIOVA',146),
        Arc('CRAIOVA','DOBRETA',120), Arc('DOBRETA','CRAIOVA',120),
        Arc('DOBRETA','MEHADIA',75), Arc('MEHADIA','DOBRETA',75),
        Arc('EFOIRE','HIRSOVA',86),  Arc('HIRSOVA','EFOIRE',86),
        Arc('FAGARAS','SIBIU',99), Arc('SIBIU','FAGARAS',99),
        Arc('HIRSOVA','URZICENI',98), Arc('URZICENI','HIRSOVA',98),
        Arc('IASI','VASLUI',92), Arc('VASLUI','IASI',92),
        Arc('IASI','NEAMT',87), Arc('NEAMT','IASI',87),
        Arc('LUGOJ','MEHADIA',70), Arc('MEHADIA','LUGOJ',70),
        Arc('LUGOJ','TIMISOARA',111), Arc('TIMISOARA','LUGOJ',111),
        Arc('ORADEA','ZERIND',71), Arc('ZERIND','ORADEA',71),
        Arc('ORADEA','SIBIU',151), Arc('SIBIU','ORADEA',151),
        Arc('PITESTI','RIMNICU VILCEA',97), Arc('RIMNICU VILCEA','PITESTI',97),
        Arc('RIMNICU VILCEA','SIBIU',80), Arc('SIBIU','RIMNICU VILCEA',80),   
        Arc('URZICENI','VASLUI',142), Arc('VASLUI','URZICENI',142)],
    start = 'DOBRETA',
    goals = {'BUCHAREST'},
    hmap = {
        'ARAD' : 366,
        'BUCHAREST' : 0,
        'CRAIOVA' : 160,
        'DOBRETA' : 242,
        'EFOIRE' : 161,
        'FAGARAS' : 178,
        'GIURGIU' : 77,
        'HIRSOVA' : 151,
        'IASI' : 226,
        'LUGOJ' : 244,
        'MEHADIA' : 241,
        'NEAMT' : 234,
        'ORADEA' : 380,
        'PITESTI' : 98,
        'RIMNICU VILCEA' :193,
        'SIBIU' : 253,
        'TIMISOARA' : 329,
        'URZICENI' : 80, 
        'VASLUI' : 199,
        'ZERIND' : 374
        }
    )


In [0]:
print("Prueba 4: Origen DOBRETA y destino BUCHAREST")
schr = AStarSearcher(problema4)
path = schr.search()
print("Ruta encontrada:", path)
print('Costo de la Ruta: ', path.cost)

Prueba 4: Origen DOBRETA y destino BUCHAREST
Costo del camino 0  + Costo de la heurística de:  DOBRETA  es:  242  -->  242
Costo del camino 75  + Costo de la heurística de:  MEHADIA  es:  241  -->  316
Costo del camino 120  + Costo de la heurística de:  CRAIOVA  es:  160  -->  280
Costo del camino 240  + Costo de la heurística de:  DOBRETA  es:  242  -->  482
Costo del camino 266  + Costo de la heurística de:  RIMNICU VILCEA  es:  193  -->  459
Costo del camino 258  + Costo de la heurística de:  PITESTI  es:  98  -->  356
Costo del camino 145  + Costo de la heurística de:  LUGOJ  es:  244  -->  389
Costo del camino 150  + Costo de la heurística de:  DOBRETA  es:  242  -->  392
Costo del camino 355  + Costo de la heurística de:  RIMNICU VILCEA  es:  193  -->  548
Costo del camino 396  + Costo de la heurística de:  CRAIOVA  es:  160  -->  556
Costo del camino 359  + Costo de la heurística de:  BUCHAREST  es:  0  -->  359
5 caminos se han ampliado y 6 rutas (paths) permanecen en la fronte

**PRUEBA 5 A***

En la prueba 5 se desea encontrar una solución: secuencia de ciudades que nos lleve desde Craiova hasta Bucarest.  Para esto se debe cambiar start = CRAIOVA en el problema de entrega, en esta prueba costo de la ruta de 239 

In [0]:
problema5 = Search_problem_from_explicit_graph(
    
   {'ARAD','BUCHAREST','CRAIOVA','DOBRETA','EFOIRE','FAGARAS','GIURGIU','HIRSOVA','IASI' ,'LUGOJ','MEHADIA','NEAMT','ORADEA','PITESTI','RIMNICU VILCEA','SIBIU',
    'TIMISOARA','URZICENI', 'VASLUI','ZERIND'},
     [  Arc('ARAD','ZERIND',75), Arc('ZERIND','ARAD',75),
        Arc('ARAD','SIBIU',140), Arc('SIBIU','ARAD',140),
        Arc('ARAD','TIMISOARA',118),Arc('TIMISOARA','ARAD',118),  
        Arc('BUCHAREST','PITESTI',101), Arc('PITESTI','BUCHAREST',101),
        Arc('BUCHAREST','FAGARAS',211), Arc('FAGARAS','BUCHAREST',211),
        Arc('BUCHAREST','URZICENI',85), Arc('URZICENI','BUCHAREST',85),
        Arc('BUCHAREST','GIURGIU',90),Arc('GIURGIU','BUCHAREST',90),
        Arc('CRAIOVA','PITESTI',138), Arc('PITESTI','CRAIOVA',138),
        Arc('CRAIOVA','RIMNICU VILCEA',146), Arc('RIMNICU VILCEA','CRAIOVA',146),
        Arc('CRAIOVA','DOBRETA',120), Arc('DOBRETA','CRAIOVA',120),
        Arc('DOBRETA','MEHADIA',75), Arc('MEHADIA','DOBRETA',75),
        Arc('EFOIRE','HIRSOVA',86),  Arc('HIRSOVA','EFOIRE',86),
        Arc('FAGARAS','SIBIU',99), Arc('SIBIU','FAGARAS',99),
        Arc('HIRSOVA','URZICENI',98), Arc('URZICENI','HIRSOVA',98),
        Arc('IASI','VASLUI',92), Arc('VASLUI','IASI',92),
        Arc('IASI','NEAMT',87), Arc('NEAMT','IASI',87),
        Arc('LUGOJ','MEHADIA',70), Arc('MEHADIA','LUGOJ',70),
        Arc('LUGOJ','TIMISOARA',111), Arc('TIMISOARA','LUGOJ',111),
        Arc('ORADEA','ZERIND',71), Arc('ZERIND','ORADEA',71),
        Arc('ORADEA','SIBIU',151), Arc('SIBIU','ORADEA',151),
        Arc('PITESTI','RIMNICU VILCEA',97), Arc('RIMNICU VILCEA','PITESTI',97),
        Arc('RIMNICU VILCEA','SIBIU',80), Arc('SIBIU','RIMNICU VILCEA',80),   
        Arc('URZICENI','VASLUI',142), Arc('VASLUI','URZICENI',142)],
    start = 'CRAIOVA',
    goals = {'BUCHAREST'},
    hmap = {
        'ARAD' : 366,
        'BUCHAREST' : 0,
        'CRAIOVA' : 160,
        'DOBRETA' : 242,
        'EFOIRE' : 161,
        'FAGARAS' : 178,
        'GIURGIU' : 77,
        'HIRSOVA' : 151,
        'IASI' : 226,
        'LUGOJ' : 244,
        'MEHADIA' : 241,
        'NEAMT' : 234,
        'ORADEA' : 380,
        'PITESTI' : 98,
        'RIMNICU VILCEA' :193,
        'SIBIU' : 253,
        'TIMISOARA' : 329,
        'URZICENI' : 80, 
        'VASLUI' : 199,
        'ZERIND' : 374
        }
    )


In [0]:
print("Prueba 5: Origen CRAIOVA y destino BUCHAREST")
schr = AStarSearcher(problema5)
path = schr.search()
print("Ruta encontrada:", path)
print('Costo de la Ruta: ', path.cost)

Prueba 5: Origen CRAIOVA y destino BUCHAREST
Costo del camino 0  + Costo de la heurística de:  CRAIOVA  es:  160  -->  160
Costo del camino 120  + Costo de la heurística de:  DOBRETA  es:  242  -->  362
Costo del camino 146  + Costo de la heurística de:  RIMNICU VILCEA  es:  193  -->  339
Costo del camino 138  + Costo de la heurística de:  PITESTI  es:  98  -->  236
Costo del camino 235  + Costo de la heurística de:  RIMNICU VILCEA  es:  193  -->  428
Costo del camino 276  + Costo de la heurística de:  CRAIOVA  es:  160  -->  436
Costo del camino 239  + Costo de la heurística de:  BUCHAREST  es:  0  -->  239
3 caminos se han ampliado y 4 rutas (paths) permanecen en la frontera
Ruta encontrada: CRAIOVA --> PITESTI --> BUCHAREST
Costo de la Ruta:  239


**PRUEBA 6 A***

En la prueba 6 se desea encontrar una solución: secuencia de ciudades que nos lleve desde Efoire hasta Bucarest.  Para esto se debe cambiar start = EFOIRE en el problema de entrega, en esta prueba costo de la ruta de 269

In [0]:
problema6 = Search_problem_from_explicit_graph(
    
   {'ARAD','BUCHAREST','CRAIOVA','DOBRETA','EFOIRE','FAGARAS','GIURGIU','HIRSOVA','IASI' ,'LUGOJ','MEHADIA','NEAMT','ORADEA','PITESTI','RIMNICU VILCEA','SIBIU',
    'TIMISOARA','URZICENI', 'VASLUI','ZERIND'},
     [  Arc('ARAD','ZERIND',75), Arc('ZERIND','ARAD',75),
        Arc('ARAD','SIBIU',140), Arc('SIBIU','ARAD',140),
        Arc('ARAD','TIMISOARA',118),Arc('TIMISOARA','ARAD',118),  
        Arc('BUCHAREST','PITESTI',101), Arc('PITESTI','BUCHAREST',101),
        Arc('BUCHAREST','FAGARAS',211), Arc('FAGARAS','BUCHAREST',211),
        Arc('BUCHAREST','URZICENI',85), Arc('URZICENI','BUCHAREST',85),
        Arc('BUCHAREST','GIURGIU',90),Arc('GIURGIU','BUCHAREST',90),
        Arc('CRAIOVA','PITESTI',138), Arc('PITESTI','CRAIOVA',138),
        Arc('CRAIOVA','RIMNICU VILCEA',146), Arc('RIMNICU VILCEA','CRAIOVA',146),
        Arc('CRAIOVA','DOBRETA',120), Arc('DOBRETA','CRAIOVA',120),
        Arc('DOBRETA','MEHADIA',75), Arc('MEHADIA','DOBRETA',75),
        Arc('EFOIRE','HIRSOVA',86),  Arc('HIRSOVA','EFOIRE',86),
        Arc('FAGARAS','SIBIU',99), Arc('SIBIU','FAGARAS',99),
        Arc('HIRSOVA','URZICENI',98), Arc('URZICENI','HIRSOVA',98),
        Arc('IASI','VASLUI',92), Arc('VASLUI','IASI',92),
        Arc('IASI','NEAMT',87), Arc('NEAMT','IASI',87),
        Arc('LUGOJ','MEHADIA',70), Arc('MEHADIA','LUGOJ',70),
        Arc('LUGOJ','TIMISOARA',111), Arc('TIMISOARA','LUGOJ',111),
        Arc('ORADEA','ZERIND',71), Arc('ZERIND','ORADEA',71),
        Arc('ORADEA','SIBIU',151), Arc('SIBIU','ORADEA',151),
        Arc('PITESTI','RIMNICU VILCEA',97), Arc('RIMNICU VILCEA','PITESTI',97),
        Arc('RIMNICU VILCEA','SIBIU',80), Arc('SIBIU','RIMNICU VILCEA',80),   
        Arc('URZICENI','VASLUI',142), Arc('VASLUI','URZICENI',142)],
    start = 'EFOIRE',
    goals = {'BUCHAREST'},
    hmap = {
        'ARAD' : 366,
        'BUCHAREST' : 0,
        'CRAIOVA' : 160,
        'DOBRETA' : 242,
        'EFOIRE' : 161,
        'FAGARAS' : 178,
        'GIURGIU' : 77,
        'HIRSOVA' : 151,
        'IASI' : 226,
        'LUGOJ' : 244,
        'MEHADIA' : 241,
        'NEAMT' : 234,
        'ORADEA' : 380,
        'PITESTI' : 98,
        'RIMNICU VILCEA' :193,
        'SIBIU' : 253,
        'TIMISOARA' : 329,
        'URZICENI' : 80, 
        'VASLUI' : 199,
        'ZERIND' : 374
        }
    )


In [0]:
print("Prueba 6: Origen EFOIRE y destino BUCHAREST")
schr = AStarSearcher(problema6)
path = schr.search()
print("Ruta encontrada:", path)
print('Costo de la Ruta: ', path.cost)

Prueba 6: Origen EFOIRE y destino BUCHAREST
Costo del camino 0  + Costo de la heurística de:  EFOIRE  es:  161  -->  161
Costo del camino 86  + Costo de la heurística de:  HIRSOVA  es:  151  -->  237
Costo del camino 184  + Costo de la heurística de:  URZICENI  es:  80  -->  264
Costo del camino 172  + Costo de la heurística de:  EFOIRE  es:  161  -->  333
Costo del camino 326  + Costo de la heurística de:  VASLUI  es:  199  -->  525
Costo del camino 282  + Costo de la heurística de:  HIRSOVA  es:  151  -->  433
Costo del camino 269  + Costo de la heurística de:  BUCHAREST  es:  0  -->  269
4 caminos se han ampliado y 3 rutas (paths) permanecen en la frontera
Ruta encontrada: EFOIRE --> HIRSOVA --> URZICENI --> BUCHAREST
Costo de la Ruta:  269


**PRUEBA 7 A***

En la prueba 7 se desea encontrar una solución: secuencia de ciudades que nos lleve desde Lugoj hasta Bucarest.  Para esto se debe cambiar start = LUGOJ en el problema de entrega, en esta prueba costo de la ruta de 504 

In [0]:
problema7 = Search_problem_from_explicit_graph(
    
   {'ARAD','BUCHAREST','CRAIOVA','DOBRETA','EFOIRE','FAGARAS','GIURGIU','HIRSOVA','IASI' ,'LUGOJ','MEHADIA','NEAMT','ORADEA','PITESTI','RIMNICU VILCEA','SIBIU',
    'TIMISOARA','URZICENI', 'VASLUI','ZERIND'},
     [  Arc('ARAD','ZERIND',75), Arc('ZERIND','ARAD',75),
        Arc('ARAD','SIBIU',140), Arc('SIBIU','ARAD',140),
        Arc('ARAD','TIMISOARA',118),Arc('TIMISOARA','ARAD',118),  
        Arc('BUCHAREST','PITESTI',101), Arc('PITESTI','BUCHAREST',101),
        Arc('BUCHAREST','FAGARAS',211), Arc('FAGARAS','BUCHAREST',211),
        Arc('BUCHAREST','URZICENI',85), Arc('URZICENI','BUCHAREST',85),
        Arc('BUCHAREST','GIURGIU',90),Arc('GIURGIU','BUCHAREST',90),
        Arc('CRAIOVA','PITESTI',138), Arc('PITESTI','CRAIOVA',138),
        Arc('CRAIOVA','RIMNICU VILCEA',146), Arc('RIMNICU VILCEA','CRAIOVA',146),
        Arc('CRAIOVA','DOBRETA',120), Arc('DOBRETA','CRAIOVA',120),
        Arc('DOBRETA','MEHADIA',75), Arc('MEHADIA','DOBRETA',75),
        Arc('EFOIRE','HIRSOVA',86),  Arc('HIRSOVA','EFOIRE',86),
        Arc('FAGARAS','SIBIU',99), Arc('SIBIU','FAGARAS',99),
        Arc('HIRSOVA','URZICENI',98), Arc('URZICENI','HIRSOVA',98),
        Arc('IASI','VASLUI',92), Arc('VASLUI','IASI',92),
        Arc('IASI','NEAMT',87), Arc('NEAMT','IASI',87),
        Arc('LUGOJ','MEHADIA',70), Arc('MEHADIA','LUGOJ',70),
        Arc('LUGOJ','TIMISOARA',111), Arc('TIMISOARA','LUGOJ',111),
        Arc('ORADEA','ZERIND',71), Arc('ZERIND','ORADEA',71),
        Arc('ORADEA','SIBIU',151), Arc('SIBIU','ORADEA',151),
        Arc('PITESTI','RIMNICU VILCEA',97), Arc('RIMNICU VILCEA','PITESTI',97),
        Arc('RIMNICU VILCEA','SIBIU',80), Arc('SIBIU','RIMNICU VILCEA',80),   
        Arc('URZICENI','VASLUI',142), Arc('VASLUI','URZICENI',142)],
    start = 'LUGOJ',
    goals = {'BUCHAREST'},
    hmap = {
        'ARAD' : 366,
        'BUCHAREST' : 0,
        'CRAIOVA' : 160,
        'DOBRETA' : 242,
        'EFOIRE' : 161,
        'FAGARAS' : 178,
        'GIURGIU' : 77,
        'HIRSOVA' : 151,
        'IASI' : 226,
        'LUGOJ' : 244,
        'MEHADIA' : 241,
        'NEAMT' : 234,
        'ORADEA' : 380,
        'PITESTI' : 98,
        'RIMNICU VILCEA' :193,
        'SIBIU' : 253,
        'TIMISOARA' : 329,
        'URZICENI' : 80, 
        'VASLUI' : 199,
        'ZERIND' : 374
        }
    )


In [0]:
print("Prueba 7: Origen LUGOJ y destino BUCHAREST")
schr = AStarSearcher(problema7)
path = schr.search()
print("Ruta encontrada:", path)
print('Costo de la Ruta: ', path.cost)

Prueba 7: Origen LUGOJ y destino BUCHAREST
Costo del camino 0  + Costo de la heurística de:  LUGOJ  es:  244  -->  244
Costo del camino 111  + Costo de la heurística de:  TIMISOARA  es:  329  -->  440
Costo del camino 70  + Costo de la heurística de:  MEHADIA  es:  241  -->  311
Costo del camino 140  + Costo de la heurística de:  LUGOJ  es:  244  -->  384
Costo del camino 145  + Costo de la heurística de:  DOBRETA  es:  242  -->  387
Costo del camino 251  + Costo de la heurística de:  TIMISOARA  es:  329  -->  580
Costo del camino 210  + Costo de la heurística de:  MEHADIA  es:  241  -->  451
Costo del camino 220  + Costo de la heurística de:  MEHADIA  es:  241  -->  461
Costo del camino 265  + Costo de la heurística de:  CRAIOVA  es:  160  -->  425
Costo del camino 385  + Costo de la heurística de:  DOBRETA  es:  242  -->  627
Costo del camino 411  + Costo de la heurística de:  RIMNICU VILCEA  es:  193  -->  604
Costo del camino 403  + Costo de la heurística de:  PITESTI  es:  98  -->

**PRUEBA 8 A***

En la prueba 8 se desea encontrar una solución: secuencia de ciudades que nos lleve desde Sibiu hasta Bucarest.  Para esto se debe cambiar start = SIBIU en el problema de entrega, en esta prueba costo de la ruta de 278 

In [0]:
problema8 = Search_problem_from_explicit_graph(
    
   {'ARAD','BUCHAREST','CRAIOVA','DOBRETA','EFOIRE','FAGARAS','GIURGIU','HIRSOVA','IASI' ,'LUGOJ','MEHADIA','NEAMT','ORADEA','PITESTI','RIMNICU VILCEA','SIBIU',
    'TIMISOARA','URZICENI', 'VASLUI','ZERIND'},
     [  Arc('ARAD','ZERIND',75), Arc('ZERIND','ARAD',75),
        Arc('ARAD','SIBIU',140), Arc('SIBIU','ARAD',140),
        Arc('ARAD','TIMISOARA',118),Arc('TIMISOARA','ARAD',118),  
        Arc('BUCHAREST','PITESTI',101), Arc('PITESTI','BUCHAREST',101),
        Arc('BUCHAREST','FAGARAS',211), Arc('FAGARAS','BUCHAREST',211),
        Arc('BUCHAREST','URZICENI',85), Arc('URZICENI','BUCHAREST',85),
        Arc('BUCHAREST','GIURGIU',90),Arc('GIURGIU','BUCHAREST',90),
        Arc('CRAIOVA','PITESTI',138), Arc('PITESTI','CRAIOVA',138),
        Arc('CRAIOVA','RIMNICU VILCEA',146), Arc('RIMNICU VILCEA','CRAIOVA',146),
        Arc('CRAIOVA','DOBRETA',120), Arc('DOBRETA','CRAIOVA',120),
        Arc('DOBRETA','MEHADIA',75), Arc('MEHADIA','DOBRETA',75),
        Arc('EFOIRE','HIRSOVA',86),  Arc('HIRSOVA','EFOIRE',86),
        Arc('FAGARAS','SIBIU',99), Arc('SIBIU','FAGARAS',99),
        Arc('HIRSOVA','URZICENI',98), Arc('URZICENI','HIRSOVA',98),
        Arc('IASI','VASLUI',92), Arc('VASLUI','IASI',92),
        Arc('IASI','NEAMT',87), Arc('NEAMT','IASI',87),
        Arc('LUGOJ','MEHADIA',70), Arc('MEHADIA','LUGOJ',70),
        Arc('LUGOJ','TIMISOARA',111), Arc('TIMISOARA','LUGOJ',111),
        Arc('ORADEA','ZERIND',71), Arc('ZERIND','ORADEA',71),
        Arc('ORADEA','SIBIU',151), Arc('SIBIU','ORADEA',151),
        Arc('PITESTI','RIMNICU VILCEA',97), Arc('RIMNICU VILCEA','PITESTI',97),
        Arc('RIMNICU VILCEA','SIBIU',80), Arc('SIBIU','RIMNICU VILCEA',80),   
        Arc('URZICENI','VASLUI',142), Arc('VASLUI','URZICENI',142)],
    start = 'SIBIU',
    goals = {'BUCHAREST'},
    hmap = {
        'ARAD' : 366,
        'BUCHAREST' : 0,
        'CRAIOVA' : 160,
        'DOBRETA' : 242,
        'EFOIRE' : 161,
        'FAGARAS' : 178,
        'GIURGIU' : 77,
        'HIRSOVA' : 151,
        'IASI' : 226,
        'LUGOJ' : 244,
        'MEHADIA' : 241,
        'NEAMT' : 234,
        'ORADEA' : 380,
        'PITESTI' : 98,
        'RIMNICU VILCEA' :193,
        'SIBIU' : 253,
        'TIMISOARA' : 329,
        'URZICENI' : 80, 
        'VASLUI' : 199,
        'ZERIND' : 374
        }
    )


In [0]:
print("Prueba 8: Origen SIBIU y destino BUCHAREST")
schr = AStarSearcher(problema8)
path = schr.search()
print("Ruta encontrada:", path)
print('Costo de la Ruta: ', path.cost)

Prueba 8: Origen SIBIU y destino BUCHAREST
Costo del camino 0  + Costo de la heurística de:  SIBIU  es:  253  -->  253
Costo del camino 80  + Costo de la heurística de:  RIMNICU VILCEA  es:  193  -->  273
Costo del camino 151  + Costo de la heurística de:  ORADEA  es:  380  -->  531
Costo del camino 99  + Costo de la heurística de:  FAGARAS  es:  178  -->  277
Costo del camino 140  + Costo de la heurística de:  ARAD  es:  366  -->  506
Costo del camino 160  + Costo de la heurística de:  SIBIU  es:  253  -->  413
Costo del camino 177  + Costo de la heurística de:  PITESTI  es:  98  -->  275
Costo del camino 226  + Costo de la heurística de:  CRAIOVA  es:  160  -->  386
Costo del camino 274  + Costo de la heurística de:  RIMNICU VILCEA  es:  193  -->  467
Costo del camino 315  + Costo de la heurística de:  CRAIOVA  es:  160  -->  475
Costo del camino 278  + Costo de la heurística de:  BUCHAREST  es:  0  -->  278
Costo del camino 198  + Costo de la heurística de:  SIBIU  es:  253  -->  45

**PRUEBA 9 A***

En la prueba 9 se desea encontrar una solución: secuencia de ciudades que nos lleve desde Timiosara hasta Bucarest.  Para esto se debe cambiar start = TIMISOARA en el problema de entrega, en esta prueba costo de la ruta de 536 

In [0]:
problema9 = Search_problem_from_explicit_graph(
    
   {'ARAD','BUCHAREST','CRAIOVA','DOBRETA','EFOIRE','FAGARAS','GIURGIU','HIRSOVA','IASI' ,'LUGOJ','MEHADIA','NEAMT','ORADEA','PITESTI','RIMNICU VILCEA','SIBIU',
    'TIMISOARA','URZICENI', 'VASLUI','ZERIND'},
     [  Arc('ARAD','ZERIND',75), Arc('ZERIND','ARAD',75),
        Arc('ARAD','SIBIU',140), Arc('SIBIU','ARAD',140),
        Arc('ARAD','TIMISOARA',118),Arc('TIMISOARA','ARAD',118),  
        Arc('BUCHAREST','PITESTI',101), Arc('PITESTI','BUCHAREST',101),
        Arc('BUCHAREST','FAGARAS',211), Arc('FAGARAS','BUCHAREST',211),
        Arc('BUCHAREST','URZICENI',85), Arc('URZICENI','BUCHAREST',85),
        Arc('BUCHAREST','GIURGIU',90),Arc('GIURGIU','BUCHAREST',90),
        Arc('CRAIOVA','PITESTI',138), Arc('PITESTI','CRAIOVA',138),
        Arc('CRAIOVA','RIMNICU VILCEA',146), Arc('RIMNICU VILCEA','CRAIOVA',146),
        Arc('CRAIOVA','DOBRETA',120), Arc('DOBRETA','CRAIOVA',120),
        Arc('DOBRETA','MEHADIA',75), Arc('MEHADIA','DOBRETA',75),
        Arc('EFOIRE','HIRSOVA',86),  Arc('HIRSOVA','EFOIRE',86),
        Arc('FAGARAS','SIBIU',99), Arc('SIBIU','FAGARAS',99),
        Arc('HIRSOVA','URZICENI',98), Arc('URZICENI','HIRSOVA',98),
        Arc('IASI','VASLUI',92), Arc('VASLUI','IASI',92),
        Arc('IASI','NEAMT',87), Arc('NEAMT','IASI',87),
        Arc('LUGOJ','MEHADIA',70), Arc('MEHADIA','LUGOJ',70),
        Arc('LUGOJ','TIMISOARA',111), Arc('TIMISOARA','LUGOJ',111),
        Arc('ORADEA','ZERIND',71), Arc('ZERIND','ORADEA',71),
        Arc('ORADEA','SIBIU',151), Arc('SIBIU','ORADEA',151),
        Arc('PITESTI','RIMNICU VILCEA',97), Arc('RIMNICU VILCEA','PITESTI',97),
        Arc('RIMNICU VILCEA','SIBIU',80), Arc('SIBIU','RIMNICU VILCEA',80),   
        Arc('URZICENI','VASLUI',142), Arc('VASLUI','URZICENI',142)],
    start = 'TIMISOARA',
    goals = {'BUCHAREST'},
    hmap = {
        'ARAD' : 366,
        'BUCHAREST' : 0,
        'CRAIOVA' : 160,
        'DOBRETA' : 242,
        'EFOIRE' : 161,
        'FAGARAS' : 178,
        'GIURGIU' : 77,
        'HIRSOVA' : 151,
        'IASI' : 226,
        'LUGOJ' : 244,
        'MEHADIA' : 241,
        'NEAMT' : 234,
        'ORADEA' : 380,
        'PITESTI' : 98,
        'RIMNICU VILCEA' :193,
        'SIBIU' : 253,
        'TIMISOARA' : 329,
        'URZICENI' : 80, 
        'VASLUI' : 199,
        'ZERIND' : 374
        }
    )


In [0]:
print("Prueba 9: Origen TIMISOARA y destino BUCHAREST")
schr = AStarSearcher(problema9)
path = schr.search()
print("Ruta encontrada:", path)
print('Costo de la Ruta: ', path.cost)

Prueba 9: Origen TIMISOARA y destino BUCHAREST
Costo del camino 0  + Costo de la heurística de:  TIMISOARA  es:  329  -->  329
Costo del camino 111  + Costo de la heurística de:  LUGOJ  es:  244  -->  355
Costo del camino 118  + Costo de la heurística de:  ARAD  es:  366  -->  484
Costo del camino 222  + Costo de la heurística de:  TIMISOARA  es:  329  -->  551
Costo del camino 181  + Costo de la heurística de:  MEHADIA  es:  241  -->  422
Costo del camino 251  + Costo de la heurística de:  LUGOJ  es:  244  -->  495
Costo del camino 256  + Costo de la heurística de:  DOBRETA  es:  242  -->  498
Costo del camino 236  + Costo de la heurística de:  TIMISOARA  es:  329  -->  565
Costo del camino 258  + Costo de la heurística de:  SIBIU  es:  253  -->  511
Costo del camino 193  + Costo de la heurística de:  ZERIND  es:  374  -->  567
Costo del camino 362  + Costo de la heurística de:  TIMISOARA  es:  329  -->  691
Costo del camino 321  + Costo de la heurística de:  MEHADIA  es:  241  -->  5

**PRUEBA 10 A***

En la prueba 10 se desea encontrar una solución: secuencia de ciudades que nos lleve desde Mehadia hasta Bucarest.  Para esto se debe cambiar start = MEHADIA en el problema de entrega, en esta prueba costo de la ruta de 434 

In [0]:
problema10 = Search_problem_from_explicit_graph(
    
   {'ARAD','BUCHAREST','CRAIOVA','DOBRETA','EFOIRE','FAGARAS','GIURGIU','HIRSOVA','IASI' ,'LUGOJ','MEHADIA','NEAMT','ORADEA','PITESTI','RIMNICU VILCEA','SIBIU',
    'TIMISOARA','URZICENI', 'VASLUI','ZERIND'},
     [  Arc('ARAD','ZERIND',75), Arc('ZERIND','ARAD',75),
        Arc('ARAD','SIBIU',140), Arc('SIBIU','ARAD',140),
        Arc('ARAD','TIMISOARA',118),Arc('TIMISOARA','ARAD',118),  
        Arc('BUCHAREST','PITESTI',101), Arc('PITESTI','BUCHAREST',101),
        Arc('BUCHAREST','FAGARAS',211), Arc('FAGARAS','BUCHAREST',211),
        Arc('BUCHAREST','URZICENI',85), Arc('URZICENI','BUCHAREST',85),
        Arc('BUCHAREST','GIURGIU',90),Arc('GIURGIU','BUCHAREST',90),
        Arc('CRAIOVA','PITESTI',138), Arc('PITESTI','CRAIOVA',138),
        Arc('CRAIOVA','RIMNICU VILCEA',146), Arc('RIMNICU VILCEA','CRAIOVA',146),
        Arc('CRAIOVA','DOBRETA',120), Arc('DOBRETA','CRAIOVA',120),
        Arc('DOBRETA','MEHADIA',75), Arc('MEHADIA','DOBRETA',75),
        Arc('EFOIRE','HIRSOVA',86),  Arc('HIRSOVA','EFOIRE',86),
        Arc('FAGARAS','SIBIU',99), Arc('SIBIU','FAGARAS',99),
        Arc('HIRSOVA','URZICENI',98), Arc('URZICENI','HIRSOVA',98),
        Arc('IASI','VASLUI',92), Arc('VASLUI','IASI',92),
        Arc('IASI','NEAMT',87), Arc('NEAMT','IASI',87),
        Arc('LUGOJ','MEHADIA',70), Arc('MEHADIA','LUGOJ',70),
        Arc('LUGOJ','TIMISOARA',111), Arc('TIMISOARA','LUGOJ',111),
        Arc('ORADEA','ZERIND',71), Arc('ZERIND','ORADEA',71),
        Arc('ORADEA','SIBIU',151), Arc('SIBIU','ORADEA',151),
        Arc('PITESTI','RIMNICU VILCEA',97), Arc('RIMNICU VILCEA','PITESTI',97),
        Arc('RIMNICU VILCEA','SIBIU',80), Arc('SIBIU','RIMNICU VILCEA',80),   
        Arc('URZICENI','VASLUI',142), Arc('VASLUI','URZICENI',142)],
    start = 'MEHADIA',
    goals = {'BUCHAREST'},
    hmap = {
        'ARAD' : 366,
        'BUCHAREST' : 0,
        'CRAIOVA' : 160,
        'DOBRETA' : 242,
        'EFOIRE' : 161,
        'FAGARAS' : 178,
        'GIURGIU' : 77,
        'HIRSOVA' : 151,
        'IASI' : 226,
        'LUGOJ' : 244,
        'MEHADIA' : 241,
        'NEAMT' : 234,
        'ORADEA' : 380,
        'PITESTI' : 98,
        'RIMNICU VILCEA' :193,
        'SIBIU' : 253,
        'TIMISOARA' : 329,
        'URZICENI' : 80, 
        'VASLUI' : 199,
        'ZERIND' : 374
        }
    )


In [0]:
print("Prueba 10: Origen MEHADIA y destino BUCHAREST")
schr = AStarSearcher(problema10)
path = schr.search()
print("Ruta encontrada:", path)
print('Costo de la Ruta: ', path.cost)


Prueba 10: Origen MEHADIA y destino BUCHAREST
Costo del camino 0  + Costo de la heurística de:  MEHADIA  es:  241  -->  241
Costo del camino 70  + Costo de la heurística de:  LUGOJ  es:  244  -->  314
Costo del camino 75  + Costo de la heurística de:  DOBRETA  es:  242  -->  317
Costo del camino 181  + Costo de la heurística de:  TIMISOARA  es:  329  -->  510
Costo del camino 140  + Costo de la heurística de:  MEHADIA  es:  241  -->  381
Costo del camino 150  + Costo de la heurística de:  MEHADIA  es:  241  -->  391
Costo del camino 195  + Costo de la heurística de:  CRAIOVA  es:  160  -->  355
Costo del camino 315  + Costo de la heurística de:  DOBRETA  es:  242  -->  557
Costo del camino 341  + Costo de la heurística de:  RIMNICU VILCEA  es:  193  -->  534
Costo del camino 333  + Costo de la heurística de:  PITESTI  es:  98  -->  431
Costo del camino 210  + Costo de la heurística de:  LUGOJ  es:  244  -->  454
Costo del camino 215  + Costo de la heurística de:  DOBRETA  es:  242  -->